{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ac017895-03b6-432c-8fa1-6c066caa01a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scipy\n",
    "import numpy as np\n",
    "# import h5py\n",
    "\n",
    "from constants import HARD_WIRED_GENOME_A_MATRIX_PATH, STRING_EXTRACTED_URL_PATH, STRING_PROTEIN_GENE_PATH, DATA_DIRECTORY, HURI_UNIQUE_GENE_SET, HARD_WIRED_GENOME_C_EDGELIST, HARD_WIRED_GENOME_B_EDGELIST, HARD_WIRED_GENOME_A_EDGELIST\n",
    "# from mapping import get_unique_genes, generate_unique_gene_set_Huri"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ec0bebb7-46f2-45bc-b80f-e28f29e249ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import scanpy as sc\n",
    "import networkx as nx\n",
    "\n",
    "from scipy.linalg import eigvalsh\n",
    "\n",
    "from anndata import AnnData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e78e88d8-37f1-40a8-a37c-79324a3c0fba",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5207330e-fddc-4512-bc60-7f441cc70253",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ordering 18582 genes\n",
      "Constructing Hard Wired Genome A Matrix\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from constants import HARD_WIRED_GENOME_A_EDGELIST, HARD_WIRED_GENOME_B_EDGELIST, HARD_WIRED_GENOME_C_EDGELIST\n",
    "\n",
    "from mapping import get_string_protein_gene_map, generate_string_mappings, generate_unique_gene_set_Huri, get_unique_genes, construct_HardWiredGenome_A_Matrix, construct_HardWiredGenome_B_Matrix, get_TF_set, construct_HardWiredGenome_A_Matrix\n",
    "\n",
    "import os\n",
    "\n",
    "from mapping import get_sorted_gene_order\n",
    "\n",
    "flat_col_order = get_sorted_gene_order()\n",
    "\n",
    "nodes = flat_col_order\n",
    "len(nodes)\n",
    "\n",
    "\n",
    "A_matrix, B_matrix, C_matrix = construct_HardWiredGenome_A_Matrix(threshold=0, nodes=nodes)\n",
    "\n",
    "A_matrix_edgelist = A_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "399af04a-5667-423e-972b-e68c98f5630b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from mapping import get_string_protein_gene_map, generate_string_mappings, generate_unique_gene_set_Huri, get_unique_genes, construct_HardWiredGenome_A_Matrix, construct_HardWiredGenome_B_Matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d43ffd74-1b09-4172-bf24-f40bbded1f7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# threshold = 150\n",
    "# # path = HARD_WIRED_GENOME_A_MATRIX_PATH.strip('.csv') + f\"_{threshold}.csv\"\n",
    "\n",
    "# path =  HARD_WIRED_GENOME_A_EDGELIST.strip('.csv') + f\"_{threshold}.csv\"\n",
    "\n",
    "# A_matrix_edgelist = pd.read_csv(path, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "572519b3-7869-41b6-89be-071ea67d1bb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracted 43945 mappings from 73467 ensembl lines\n",
      "Extracted 38606 mappings from gene id gtf file\n",
      "Total Mappings Extracted 59979\n"
     ]
    }
   ],
   "source": [
    "from network_utils import generate_gene_id_name_map, get_master_regulator_list\n",
    "\n",
    "gene_id_name_map, gene_name_id_map = generate_gene_id_name_map()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e3a5b100-5587-405f-8450-98dd6bbcba87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracted 43945 mappings from 73467 ensembl lines\n",
      "Extracted 38606 mappings from gene id gtf file\n",
      "Total Mappings Extracted 59979\n",
      "Number of Master regulators: 617\n",
      "total master regulators from Old list: 439\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "688"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from mapping import load_master_regulators, get_master_regulator_list, get_TF_lists\n",
    "\n",
    "master_regulators_new = load_master_regulators()\n",
    "master_regulator_list = get_master_regulator_list()\n",
    "\n",
    "set(master_regulators_new).intersection(set(master_regulator_list))\n",
    "\n",
    "len(set(master_regulators_new).union(set(master_regulator_list)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e9b6c701-b184-46f0-840d-80550a1c7662",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "367"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(master_regulators_new).intersection(set(master_regulator_list)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "875bc2d9-9cd7-4a91-8194-133fec9f8103",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Total Activators : 174\n",
      "Total Reprossors : 705\n",
      "Total Conflicted : 729\n",
      "Total Transcription Factors : 1608\n",
      "Total TFs identified: 1857\n"
     ]
    }
   ],
   "source": [
    "from mapping import load_master_regulators, get_master_regulator_list, get_TF_lists\n",
    "repressorlist, activatorlist, conflictedlist, tf_list = get_TF_lists()\n",
    "\n",
    "print(f'Total TFs identified: {len(set(tf_list).union(set(master_regulators_new)))}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d334182d-c815-434d-a1f9-3b173f0365d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processed 1639 Transcription Factors\n",
      "2       Activation  Repression  Unknown\n",
      "0                                      \n",
      "AATF             1           1        3\n",
      "ABL1             8           3        1\n",
      "AES              2           0        1\n",
      "AHR              8           2       16\n",
      "AIP              1           4        0\n",
      "...            ...         ...      ...\n",
      "ZNF410           0           0        1\n",
      "ZNF423           0           0        1\n",
      "ZNF444           1           0        0\n",
      "ZNF76            1           1        1\n",
      "ZNRD1            2           0        0\n",
      "\n",
      "[795 rows x 3 columns]\n",
      "          0       1           2                  3\n",
      "0      AATF     BAX  Repression           22909821\n",
      "1      AATF  CDKN1A     Unknown           17157788\n",
      "2      AATF    KLK3     Unknown           23146908\n",
      "3      AATF     MYC  Activation           20549547\n",
      "4      AATF    TP53     Unknown           17157788\n",
      "...     ...     ...         ...                ...\n",
      "9391  ZNF76  CDKN1A  Repression           15280358\n",
      "9392  ZNF76  PCYT1A  Activation           14702349\n",
      "9393  ZNF76  TALDO1     Unknown           14702349\n",
      "9394  ZNRD1   ABCB1  Activation           16373708\n",
      "9395  ZNRD1    BCL2  Activation  16373708;16609701\n",
      "\n",
      "[9396 rows x 4 columns]\n",
      " Total Activators : 3149\n",
      "Total Reprossors : 1922\n",
      "Total Conflicted : 4325\n",
      "Total Transcription Factors : 1639\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from constants import HUMAN_TF_ACTIVITY_PATH\n",
    "\n",
    "tflist = get_TF_set()\n",
    "TF_activity_TRRUST = pd.read_csv(HUMAN_TF_ACTIVITY_PATH, sep='\\t', header=None)\n",
    "\n",
    "activity_scores = TF_activity_TRRUST.groupby(0)[2].value_counts().unstack(fill_value=0)\n",
    "print(activity_scores)\n",
    "\n",
    "print(TF_activity_TRRUST)\n",
    "\n",
    "\n",
    "log = True\n",
    "\n",
    "if log:\n",
    "    print(f' Total Activators : {len(activatorlist)}')\n",
    "    print(f'Total Reprossors : {len(repressorlist)}')\n",
    "    print(f'Total Conflicted : {len(conflictedlist)}')\n",
    "\n",
    "    print(f'Total Transcription Factors : {len(tflist)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a9f6c9c8-f399-4b11-97c2-0041dcf949d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Master Regulators : 560\n",
      "Total intersection with Transcription Factors 367\n"
     ]
    }
   ],
   "source": [
    "print(f\"Total Master Regulators : {len(set(master_regulators_new).intersection(set(genes)))}\")\n",
    "print(f\"Total intersection with Transcription Factors {len(set(master_regulators_new).intersection(set(tflist)))}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "1f019e7c-4370-41ae-9d0f-8c52a2f2692f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/nuvi/.local/lib/python3.11/site-packages/anndata/_core/anndata.py:740: UserWarning: \n",
      "AnnData expects .obs.index to contain strings, but got values like:\n",
      "    [7, 11, 12]\n",
      "\n",
      "    Inferred to be: integer\n",
      "\n",
      "  value_idx = self._prep_dim_index(value.index, attr)\n"
     ]
    }
   ],
   "source": [
    "import h5py\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from anndata import AnnData\n",
    "import scanpy as sc\n",
    "\n",
    "RNASEQ_PATH = '/scratch/indikar_root/indikar1/shared_data/HWG/data/RNAseq'\n",
    "\n",
    "# mat_file = h5py.File('/Users/ramprakash/development/lab_projects/Rajapakse_lab/data/HWG/RNAseq/RNAseq Vectors 2015-2024.mat', 'r')\n",
    "# print(mat_file)\n",
    "\n",
    "X = pd.read_csv(f'{RNASEQ_PATH}/X.csv')\n",
    "X_arr = X.to_numpy()\n",
    "obs = pd.read_csv(f'{RNASEQ_PATH}/obs.csv')\n",
    "var = pd.read_csv(f'{RNASEQ_PATH}/var.csv')\n",
    "\n",
    "\n",
    "\n",
    "fibroblast_expression = X_arr[7]\n",
    "fibroblast_expression = np.nan_to_num(fibroblast_expression, nan=0)\n",
    "esc_expression = X_arr[11]\n",
    "esc_expression = np.nan_to_num(esc_expression, nan=0)\n",
    "myotube_expression = X_arr[12]\n",
    "myotube_expression = np.nan_to_num(myotube_expression, nan=0)\n",
    "\n",
    "fibroblast_nz_indices = np.nonzero(fibroblast_expression)\n",
    "esc_nz_indices = np.nonzero(esc_expression)\n",
    "myotube_nz_indices = np.nonzero(myotube_expression)\n",
    "\n",
    "obs = obs.iloc[[7, 11, 12]]\n",
    "samples = ['Fibroblast', 'EmbryonicStemCells', 'Myotube']\n",
    "\n",
    "\n",
    "obs['CellType'] = samples\n",
    "\n",
    "X = np.vstack([fibroblast_expression, esc_expression, myotube_expression])\n",
    "\n",
    "adata = AnnData(X=X)\n",
    "\n",
    "adata.var = var        # columns (genes)\n",
    "adata.var_names = var['GeneStableID'].astype(str)\n",
    "adata.var = adata.var.drop(columns=['GeneStableID'])\n",
    "\n",
    "adata.obs = obs\n",
    "adata.obs_names = adata.obs[\"CellType\"].astype(str)\n",
    "adata.obs = adata.obs.drop(columns=['CellType'])\n",
    "\n",
    "\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "\n",
    "adata.write('./data/Genexpression.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c546821c-f249-44e5-a7d3-89fa64e66ea7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracted 43945 mappings from 73467 ensembl lines\n",
      "Extracted 38606 mappings from gene id gtf file\n",
      "Total Mappings Extracted 59979\n"
     ]
    }
   ],
   "source": [
    "from mapping import generate_gene_id_name_map \n",
    "\n",
    "gene_id_name_map, gene_name_id_map = generate_gene_id_name_map()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "7dbe3c00-1048-477e-82e5-c0d5ca468b9f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    gene_name    cell_type  is_tf   source    category\n",
      "697   RARRES2  fibroblasts  False  panglao  fibroblast\n",
      "698     CELA1  fibroblasts  False  panglao  fibroblast\n",
      "699       LUM  fibroblasts  False  panglao  fibroblast\n",
      "700     PRRX1  fibroblasts   True  panglao  fibroblast\n",
      "701    SCARA5  fibroblasts  False  panglao  fibroblast\n",
      "..        ...          ...    ...      ...         ...\n",
      "924    COL8A1  fibroblasts  False  panglao  fibroblast\n",
      "925      PTX3  fibroblasts  False  panglao  fibroblast\n",
      "926     CEBPB  fibroblasts   True  panglao  fibroblast\n",
      "927      PKD2  fibroblasts  False  panglao  fibroblast\n",
      "928    CTHRC1  fibroblasts  False  panglao  fibroblast\n",
      "\n",
      "[232 rows x 5 columns]\n",
      "230\n"
     ]
    }
   ],
   "source": [
    "marker_gene_df = pd.read_csv('./data/clean_marker_genes.csv')\n",
    "\n",
    "fibroblast_markers = marker_gene_df[marker_gene_df['cell_type'] == 'fibroblasts']\n",
    "print(fibroblast_markers)\n",
    "marker_genes = list(fibroblast_markers['gene_name'])\n",
    "\n",
    "gene_id_name_map, gene_name_id_map = generate_gene_id_name_map()\n",
    "\n",
    "marker_ids = [ gene_name_id_map[i] for i in marker_genes ]\n",
    "\n",
    "expr_values = adata.X[adata.obs_names == 'Fibroblast'].flatten()\n",
    "\n",
    "expr_series = pd.Series(expr_values, index=adata.var_names)\n",
    "marker_expr = expr_series[expr_series.index.isin(marker_ids)]\n",
    "print(len(marker_expr))\n",
    "\n",
    "\n",
    "# threshold = marker_expr.mean()\n",
    "threshold = marker_expr.median()\n",
    "high_expr = expr_series[expr_series > threshold]\n",
    "\n",
    "top_genes = high_expr.sort_values(ascending=False)\n",
    "\n",
    "top_genes = top_genes[top_genes.index.notna()]\n",
    "\n",
    "fibroblast_expressed_genes = top_genes.index.tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "4935fe04-961b-450d-a269-f60794d18255",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2053901/452622973.py:66: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  ts_repression_edges = concat_df[\n",
      "/tmp/ipykernel_2053901/452622973.py:70: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  st_repression_edges = concat_df[\n"
     ]
    }
   ],
   "source": [
    "fibroblast_filtered_edgelist = A_matrix_edgelist[\n",
    "    A_matrix_edgelist['source'].isin(fibroblast_expressed_genes) & A_matrix_edgelist['target'].isin(fibroblast_expressed_genes)\n",
    "]\n",
    "\n",
    "current_edgelist = fibroblast_filtered_edgelist\n",
    "\n",
    "start_index = 0\n",
    "end_index = 18581\n",
    "\n",
    "# ENSG00000129152 - MYOD1\n",
    "perturbed_edges = ['ENSG00000129152']\n",
    "move_list = ['ENSG00000129152']\n",
    "\n",
    "# filtered_edgelist, filtered_nodes = get_piece_network(current_edgelist, A_matrix_edgelist, start_index, end_index, flat_col_order, perturbed_edges)\n",
    "filtered_edgelist, filtered_nodes, normal_edges, activator_edges, repressor_edges = get_piece_network(current_edgelist, A_matrix_edgelist, start_index, end_index, flat_col_order, perturbed_edges)\n",
    "# filtered_graph = construct_graph(filtered_edgelist, filtered_nodes, normal_edges, activator_edges, repressor_edges, perturbed_edges, node_attrs, time_based=False, move_list=move_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "bd7d31fb-deef-4a6c-b553-1864ee9b6a1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Fibroblast expressed genes 5006\n",
      "Number of ESC expressed genes 7018\n",
      "Number of Myotube expressed genes 7478\n"
     ]
    }
   ],
   "source": [
    "gene_names = adata.var_names\n",
    "cell_names = adata.obs_names\n",
    "\n",
    "gen_threshold = 0.2\n",
    "\n",
    "X_dense = adata.X.toarray() if not isinstance(adata.X, np.ndarray) else adata.X\n",
    "\n",
    "\n",
    "expressed_genes_per_cell = []\n",
    "\n",
    "for i, row in enumerate(X_dense):\n",
    "    expressed = gene_names[row > gen_threshold]\n",
    "    expressed_genes_per_cell.append(list(expressed))\n",
    "\n",
    "# Optional: convert to DataFrame for easier inspection\n",
    "df = pd.DataFrame({\n",
    "    'cell': cell_names,\n",
    "    'expressed_genes': expressed_genes_per_cell\n",
    "})\n",
    "\n",
    "print(\"Number of Fibroblast expressed genes\", len(df.iloc[0].expressed_genes))\n",
    "print(\"Number of ESC expressed genes\", len(df.iloc[1].expressed_genes))\n",
    "print(\"Number of Myotube expressed genes\", len(df.iloc[2].expressed_genes))\n",
    "\n",
    "fibroblast_expressed_genes = df.iloc[0].expressed_genes\n",
    "esc_expressed_genes = df.iloc[1].expressed_genes\n",
    "myotube_expressed_genes = df.iloc[2].expressed_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a40ed82b-e212-4318-8f01-4d6969b254d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_A_matrix_edgelist = A_matrix_edgelist[\n",
    "    A_matrix_edgelist['source'].isin(flat_col_order) & A_matrix_edgelist['target'].isin(flat_col_order)\n",
    "]\n",
    "\n",
    "fil_path =  HARD_WIRED_GENOME_A_EDGELIST.strip('.csv') + f\"filetered_{threshold}.csv\"\n",
    "\n",
    "filtered_A_matrix_edgelist.to_csv(fil_path, index=False)\n",
    "\n",
    "\n",
    "A_matrix_edgelist = filtered_A_matrix_edgelist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7cff6e85-d81c-45d1-ae1d-08287ceb561b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fibroblast_expressed_genes\n",
    "# myotube_expressed_genes\n",
    "# esc_expressed_genes\n",
    "\n",
    "fibroblast_filtered_edgelist = A_matrix_edgelist[\n",
    "    A_matrix_edgelist['source'].isin(fibroblast_expressed_genes) & A_matrix_edgelist['target'].isin(fibroblast_expressed_genes)\n",
    "]\n",
    "myotube_filtered_edgelist = A_matrix_edgelist[\n",
    "    A_matrix_edgelist['source'].isin(myotube_expressed_genes) & A_matrix_edgelist['target'].isin(myotube_expressed_genes)\n",
    "]\n",
    "esc_filtered_edgelist = A_matrix_edgelist[\n",
    "    A_matrix_edgelist['source'].isin(esc_expressed_genes) & A_matrix_edgelist['target'].isin(esc_expressed_genes)\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2c4fa21f-5c3a-47c1-a9e9-23d517721274",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype = 'Fibroblast'\n",
    "threshold = 150\n",
    "\n",
    "\n",
    "# path = HARD_WIRED_GENOME_A_MATRIX_PATH.strip('.csv') + f\"_{threshold}.csv\"\n",
    "\n",
    "path =  HARD_WIRED_GENOME_A_EDGELIST.strip('.csv') + f\"_{threshold}_{celltype}_{gen_threshold}.csv\"\n",
    "\n",
    "# fibroblast_filtered_edgelist.to_csv(path, index=False)\n",
    "fibroblast_filtered_edgelist = pd.read_csv(path)\n",
    "\n",
    "\n",
    "\n",
    "celltype = 'Myotube'\n",
    "threshold = 150\n",
    "\n",
    "path =  HARD_WIRED_GENOME_A_EDGELIST.strip('.csv') + f\"_{threshold}_{celltype}_{gen_threshold}.csv\"\n",
    "\n",
    "# df = pd.read_csv(path, index_col=0)\n",
    "# myotube_filtered_edgelist.to_csv(path, index=False)\n",
    "\n",
    "myotube_filtered_edgelist = pd.read_csv(path)\n",
    "\n",
    "\n",
    "\n",
    "celltype = 'EmbryonicStemCells'\n",
    "threshold = 150\n",
    "\n",
    "path =  HARD_WIRED_GENOME_A_EDGELIST.strip('.csv') + f\"_{threshold}_{celltype}_{gen_threshold}.csv\"\n",
    "\n",
    "# df = pd.read_csv(path, index_col=0)\n",
    "# esc_filtered_edgelist.to_csv(path, index=False)\n",
    "\n",
    "esc_filtered_edgelist = pd.read_csv(path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90499384-91d8-404c-9793-f43927f071d6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "7b1aa19d-d66b-4b6e-b2f7-50b5586670be",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "HARD_WIRED_GENOME_BP_MATRIX = f\"{DATA_DIRECTORY}/HWG/BP_Matrix.csv\"\n",
    "\n",
    "c_matrix = pd.read_csv(HARD_WIRED_GENOME_C_EDGELIST)\n",
    "\n",
    "master_regulators = c_matrix[c_matrix['source'] == c_matrix['target']]\n",
    "\n",
    "master_regulator_list = list(c_matrix['source'])\n",
    "\n",
    "transcription_factor_matx = c_matrix[c_matrix['source'] != c_matrix['target']]\n",
    "\n",
    "transcription_factor_list = pd.unique(transcription_factor_matx[['source', 'target']].values.ravel())\n",
    "\n",
    "BP_matrix = pd.read_csv(HARD_WIRED_GENOME_BP_MATRIX, index_col=0)\n",
    "\n",
    "column_means = BP_matrix.mean()\n",
    "\n",
    "sign_dict = column_means.apply(lambda x: 1 if x > 0 else -1 if x < 0 else 0).to_dict()\n",
    "\n",
    "gene_id_name_map, name_id_map = generate_gene_id_name_map()\n",
    "\n",
    "# POU5F1, SOX2, KLF4, MYC\n",
    "gene_id_name_map.update(dict(zip( ['ENSG00000134323', 'ENSG00000111704', 'ENSG00000204531', 'ENSG00000129152', 'ENSG00000136826', 'ENSG00000136997', 'ENSG00000181449', 'ENSG00000122180'], ['MYCN', 'NANOG' ,'POU5F1', 'MYOD1', 'KLF4', 'MYC', 'SOX2', 'MYOG'])))\n",
    "\n",
    "# gene_id_name_map\n",
    "\n",
    "def get_piece_network(edgelist, full_edgelist, start_index, end_index, order, perturbed_edges=[]):\n",
    "\n",
    "    global sign_dict\n",
    "\n",
    "    list_chunk = order[start_index:end_index]\n",
    "\n",
    "    normal_edges = []\n",
    "    activator_edges = []\n",
    "    repressor_edges = []\n",
    "    for edge in perturbed_edges:\n",
    "        sign = sign_dict.get(edge, 0)\n",
    "        if sign == -1:\n",
    "            repressor_edges.append(edge)\n",
    "        elif sign == 1:\n",
    "            activator_edges.append(edge)\n",
    "        else:\n",
    "            normal_edges.append(edge)\n",
    "\n",
    "\n",
    "    # get only the activators from the full edgelist\n",
    "\n",
    "    ts_perturbed_edgelist = full_edgelist[\n",
    "            full_edgelist['source'].isin(activator_edges) & full_edgelist['target'].isin(list_chunk)\n",
    "    ]\n",
    "\n",
    "    st_perturbed_edgelist = full_edgelist[\n",
    "            full_edgelist['source'].isin(list_chunk) & full_edgelist['target'].isin(activator_edges)\n",
    "    ]\n",
    "\n",
    "    perturbed_edgelist = pd.concat([ts_perturbed_edgelist, st_perturbed_edgelist], ignore_index=True)\n",
    "    \n",
    "\n",
    "    filtered_edgelist = edgelist[\n",
    "            edgelist['source'].isin(list_chunk) & edgelist['target'].isin(list_chunk)\n",
    "    ]\n",
    "\n",
    "    concat_df = pd.concat([perturbed_edgelist, filtered_edgelist], ignore_index=True)\n",
    "\n",
    "\n",
    "    \n",
    "    ts_repression_edges = concat_df[\n",
    "        concat_df['source'].isin(repressor_edges) & filtered_edgelist['target'].isin(activator_edges)\n",
    "    ]\n",
    "    \n",
    "    st_repression_edges = concat_df[\n",
    "        concat_df['source'].isin(activator_edges) & filtered_edgelist['target'].isin(repressor_edges)\n",
    "    ]\n",
    "\n",
    "    repression_edgelist = pd.concat([ts_repression_edges, st_repression_edges], ignore_index=True)\n",
    "\n",
    "    concat_df = concat_df[~concat_df['source'].isin(repressor_edges)]\n",
    "    concat_df = concat_df[~concat_df['target'].isin(repressor_edges)]\n",
    "\n",
    "    fin_df = pd.concat([concat_df, repression_edgelist], ignore_index=True)\n",
    "\n",
    "    return fin_df, list_chunk + perturbed_edges, normal_edges, activator_edges, repressor_edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95f36ef8-13a4-4c79-bb32-2a2e5adb292c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def construct_graph(filtered_edgelist, filtered_nodes, perturbed_edges, normal_edges, activator_edges, repressor_edges, node_attrs, time_based=False, move_list=[]):\n",
    "    global BP_matrix\n",
    "    global transcription_factor_list\n",
    "    global master_regulator_list\n",
    "    global gene_id_name_map\n",
    "    column_means = BP_matrix.mean()\n",
    "\n",
    "    sign_dict = column_means.apply(lambda x: 1 if x >= 0 else -1 if x < 0 else 0).to_dict()\n",
    "\n",
    "\n",
    "    highlight_color = \"#a345f0\"\n",
    "    repressor_color = \"#d77132\"\n",
    "    initiator_color = \"#f83e56\"\n",
    "\n",
    "    # g1_color = '#0096c7'\n",
    "    # g2_color = '#00b4d8'\n",
    "    # g3_color = \"#90e0ef\"\n",
    "    g1_color = \"#6b7fa6\"\n",
    "    g2_color = \"#86c9d7\"\n",
    "    g3_color = \"#c6d6d8\"\n",
    "    filtered_graph = nx.from_pandas_edgelist(filtered_edgelist, 'source', 'target', edge_attr='weight', create_using=nx.Graph())\n",
    "    filtered_graph.add_nodes_from(filtered_nodes)\n",
    "    filtered_graph.add_nodes_from(activator_edges)\n",
    "    filtered_graph.add_nodes_from(repressor_edges)\n",
    "    filtered_graph.add_nodes_from(move_list)\n",
    "\n",
    "\n",
    "\n",
    "    # filtered_graph = connect_unconnected_nodes(filtered_graph)\n",
    "\n",
    "\n",
    "    centrality = nx.degree_centrality(filtered_graph)\n",
    "    cent_values = np.array(list(centrality.values()))\n",
    "    cent_min, cent_max = cent_values.min(), cent_values.max()\n",
    "    norm_centrality = {node: (val - cent_min) / (cent_max - cent_min) for node, val in centrality.items()}\n",
    "\n",
    "    # Use a colormap (e.g., 'viridis') to map centrality → color\n",
    "    # cmap = cm.get_cmap('winter')\n",
    "    cmap = custom_cmap\n",
    "    node_colors = {node: cmap(norm_centrality[node]) for node in filtered_graph.nodes()}\n",
    "\n",
    "    degree_centrality = nx.degree_centrality(filtered_graph)\n",
    "\n",
    "\n",
    "    # for dynamic graphs\n",
    "    timepoints = len(perturbed_edges) + 1\n",
    "\n",
    "    for node in node_attrs:\n",
    "        key = node['key']\n",
    "        nat = node['attributes']\n",
    "        x = nat['x']\n",
    "        y = nat['y']\n",
    "        if key in filtered_graph:\n",
    "            # degree = degree_centrality[key]\n",
    "            degree = filtered_graph.degree[key]\n",
    "            filtered_graph.nodes[key]['size'] = 15 + 1 * degree\n",
    "\n",
    "            if abs(x) > 3000 or abs(y) > 3000:\n",
    "                filtered_graph.nodes[key]['x'] = 0\n",
    "                filtered_graph.nodes[key]['y'] = 0\n",
    "\n",
    "            filtered_graph.nodes[key]['x'] = x\n",
    "            filtered_graph.nodes[key]['y'] = y\n",
    "            if time_based:\n",
    "                filtered_graph.nodes[key]['color_dynamic'] = f\"<[{0},{timepoints}{'#%02x%02x%02x' % tuple(int(255 * c) for c in node_colors[key][:3])}]>\"\n",
    "                filtered_graph.nodes[key]['perturbed_dynamic'] = f\"<[{0},{timepoints}{'#%02x%02x%02x' % tuple(int(255 * c) for c in node_colors[key][:3])}]>\"\n",
    "                filtered_graph.nodes[key]['color'] = '#%02x%02x%02x' % tuple(int(255 * c) for c in node_colors[key][:3])\n",
    "            else:\n",
    "                filtered_graph.nodes[key]['color'] = '#%02x%02x%02x' % tuple(int(255 * c) for c in node_colors[key][:3])\n",
    "\n",
    "            if key in master_regulator_list:\n",
    "                filtered_graph.nodes[key]['color'] = g1_color\n",
    "                filtered_graph.nodes[key]['type'] = 'g1'\n",
    "            elif key in transcription_factor_list:\n",
    "                filtered_graph.nodes[key]['color'] = g2_color\n",
    "                filtered_graph.nodes[key]['type'] = 'g2'\n",
    "            else:\n",
    "                filtered_graph.nodes[key]['color'] = g3_color\n",
    "                filtered_graph.nodes[key]['type'] = 'g3'\n",
    "\n",
    "\n",
    "\n",
    "            filtered_graph.nodes[key]['genename'] = gene_id_name_map.get(key, \"NA\")\n",
    "            filtered_graph.nodes[key]['start'] = 0\n",
    "            filtered_graph.nodes[key]['end'] = timepoints + 10\n",
    "            filtered_graph.nodes[key]['level'] = 1\n",
    "\n",
    "            filtered_graph.nodes[key]['interval'] = f\"<[{0},{timepoints + 10}]>\"\n",
    "\n",
    "            if not(filtered_graph.nodes[key].get('perturbed')):\n",
    "                filtered_graph.nodes[key]['perturbed'] = False\n",
    "\n",
    "\n",
    "    splits = 10\n",
    "    if timepoints > splits:\n",
    "        partitions = timepoints // splits\n",
    "    else:\n",
    "        partitions = 2\n",
    "    count = 0\n",
    "    for key in perturbed_edges:\n",
    "        if key in filtered_graph:\n",
    "            if key in normal_edges:\n",
    "                filtered_graph.nodes[key]['color'] = highlight_color\n",
    "                filtered_graph.nodes[key]['level'] = 2\n",
    "            elif key in activator_edges:\n",
    "                filtered_graph.nodes[key]['color'] = highlight_color\n",
    "                filtered_graph.nodes[key]['level'] = 3\n",
    "            else:\n",
    "                filtered_graph.nodes[key]['color'] = repressor_color\n",
    "                filtered_graph.nodes[key]['level'] = 4\n",
    "                \n",
    "\n",
    "            \n",
    "            if time_based and not(filtered_graph.nodes[key].get('perturbed')):\n",
    "                filtered_graph.nodes[key]['color_dynamic'] = f\"<[{0},{count},{'#%02x%02x%02x' % tuple(int(255 * c) for c in node_colors[key][:3])}],[{count}, {timepoints},{highlight_color}]>\"\n",
    "                filtered_graph.nodes[key]['perturbed_dynamic'] = f\"<[{0},{count},{'#%02x%02x%02x' % tuple(int(255 * c) for c in node_colors[key][:3])}],[{count}, {timepoints},{highlight_color}]>\"\n",
    "                filtered_graph.nodes[key]['color'] = highlight_color\n",
    "            else:\n",
    "                if sign_dict.get(key) == 1:\n",
    "                    filtered_graph.nodes[key]['color'] = highlight_color\n",
    "                else:\n",
    "                    filtered_graph.nodes[key]['color'] = repressor_color\n",
    "\n",
    "            if not(filtered_graph.nodes[key].get('perturbed')):\n",
    "                filtered_graph.nodes[key]['perturbed'] = True\n",
    "\n",
    "        count += 1\n",
    "\n",
    "    ncount = 0\n",
    "    for nn in move_list:\n",
    "        if nn in filtered_graph:\n",
    "            filtered_graph.nodes[nn]['x'] = 2000 + (ncount * 10)\n",
    "            filtered_graph.nodes[nn]['y'] = 2000 + (ncount * 10)\n",
    "            if nn in filtered_graph:\n",
    "                filtered_graph.nodes[nn]['color'] = initiator_color\n",
    "            \n",
    "            filtered_graph.nodes[nn]['size'] = max(filtered_graph.nodes[nn].get('size', 0), 50)\n",
    "\n",
    "            if not(filtered_graph.nodes[nn].get('perturbed')):\n",
    "                filtered_graph.nodes[nn]['perturbed'] = True\n",
    "\n",
    "        ncount += 1\n",
    "\n",
    "        # if count % partitions == 0:\n",
    "        #     nx.write_graphml(filtered_graph, f\"{savepath}/final_graphs/combination_perturbed_third_order_time_based_tp_{count}.graphml\")\n",
    "\n",
    "\n",
    "    return filtered_graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47190017-32eb-4c49-8607-2b377ab4085a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx\n",
    "import json\n",
    "\n",
    "def export_graph_as_json(G, file_path):\n",
    "    ll = [ ]\n",
    "    for n in G.nodes():\n",
    "        nodeattrs = G.nodes[n]\n",
    "        genename = str(nodeattrs.get('genename', \"NA\"))\n",
    "        if 'genename' in nodeattrs:\n",
    "            del nodeattrs['genename']\n",
    "        if genename == \"NA\" or None:\n",
    "            genename = gene_id_name_map.get(str(n), str(n))\n",
    "        ll.append({\"data\": {\"id\": str(n), \"genename\": genename, **nodeattrs}})\n",
    "\n",
    "    # Convert to Cytoscape.js format\n",
    "    cy_data = {\n",
    "        \"nodes\": ll,\n",
    "        \"edges\": [\n",
    "            {\"data\": {\"id\": f\"{u}-{v}\", \"source\": str(u), \"target\": str(v), **G.edges[u, v]}} for u, v in G.edges()\n",
    "        ]\n",
    "    }\n",
    "\n",
    "    # Save to JSON\n",
    "    with open(file_path, \"w\") as f:\n",
    "        json.dump(cy_data, f, indent=2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f47fd2a6-4467-4bd6-a9ba-581f85acac47",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltype = 'all_cells'\n",
    "current_edgelist = A_matrix_edgelist\n",
    "savepath = f'{HWG_BASE_PATH}/combinations/{celltype}'\n",
    "\n",
    "start_index = 9000\n",
    "end_index = 10000\n",
    "\n",
    "perturbed_edges=[]\n",
    "node_attrs=[]\n",
    "move_list = []\n",
    "\n",
    "# filtered_edgelist, filtered_nodes = get_piece_network(current_edgelist, A_matrix_edgelist, start_index, end_index, flat_col_order, perturbed_edges)\n",
    "filtered_edgelist, filtered_nodes, normal_edges, activator_edges, repressor_edges = get_piece_network(current_edgelist, A_matrix_edgelist, start_index, end_index, flat_col_order, perturbed_edges)\n",
    "filtered_graph = construct_graph(filtered_edgelist, filtered_nodes, normal_edges, activator_edges, repressor_edges, perturbed_edges, node_attrs, time_based=False, move_list=move_list)\n",
    "\n",
    "nx.write_graphml(filtered_graph, f\"{savepath}/combination_perturbed_{threshold}_{celltype}_start_end_{start_index}_{end_index}.graphml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea2fb3c5-4f3b-43c7-b707-1ab7a0245c9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# \n",
    "# celltype = 'fibroblast'\n",
    "celltype = 'myotube'\n",
    "# celltype = 'esc'\n",
    "# celltype = 'all_cells'\n",
    "# current_edgelist = A_matrix_edgelist\n",
    "# current_edgelist = fibroblast_filtered_edgelist\n",
    "current_edgelist = myotube_filtered_edgelist\n",
    "# current_edgelist = esc_filtered_edgelist\n",
    "\n",
    "savepath = f'{HWG_BASE_PATH}/combinations/{celltype}'\n",
    "graph_json_path = './visualization/data/'\n",
    "\n",
    "start_index = 6000\n",
    "end_index = 7000\n",
    "time_based = False\n",
    "\n",
    "import json\n",
    "# with open('/Users/ramprakash/development/lab_projects/Rajapakse_lab/projects/HardWiredGenome/graphs/connected_A_matrix.json') as f:\n",
    "# with open(f'/Users/ramprakash/development/lab_projects/Rajapakse_lab/projects/HardWiredGenome/graphs/combination_perturbed_150_all_cells_{start_index}_{end_index}.json') as f:\n",
    "# with open(f'{HWG_BASE_PATH}/graphs/combination_perturbed_150_all_cells_{start_index}_{end_index}.json') as f:\n",
    "with open(f'./graphs/combination_perturbed_150_all_cells_{start_index}_{end_index}.json') as f:\n",
    "# with open(f'./graphs/combination_perturbed_150_all_cells_{6000}_{7000}.json') as f:\n",
    "    data = json.load(f)\n",
    "node_attrs = data['nodes']\n",
    "\n",
    "\n",
    "\n",
    "perturbed_edges = []\n",
    "move_list = []\n",
    "\n",
    "\n",
    "# base network\n",
    "\n",
    "filtered_edgelist, filtered_nodes, normal_edges, activator_edges, repressor_edges = get_piece_network(current_edgelist, A_matrix_edgelist, start_index, end_index, flat_col_order, [])\n",
    "filtered_graph = construct_graph(filtered_edgelist, filtered_nodes, normal_edges, activator_edges, repressor_edges, perturbed_edges, node_attrs, time_based=time_based, move_list=move_list)\n",
    "\n",
    "nx.write_graphml(filtered_graph, f\"{savepath}/combination_perturbed_base_time_based_{threshold}_{celltype}_start_end_{start_index}_{end_index}.graphml\")\n",
    "\n",
    "export_graph_as_json(filtered_graph, f\"{graph_json_path}/combination_perturbed_base_time_based_{threshold}_{celltype}_start_end_{start_index}_{end_index}.json\")\n",
    "\n",
    "base_myotube_network = filtered_graph.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a60e401-df9f-4ea1-bff1-33c53d602cae",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "import matplotlib.cm as cm\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "celltype = 'fibroblast'\n",
    "# celltype = 'myotube'\n",
    "# celltype = 'esc'\n",
    "# celltype = 'all_cells'\n",
    "# current_edgelist = A_matrix_edgelist\n",
    "current_edgelist = fibroblast_filtered_edgelist\n",
    "# current_edgelist = myotube_filtered_edgelist\n",
    "# current_edgelist = esc_filtered_edgelist\n",
    "\n",
    "savepath = f'{HWG_BASE_PATH}/combinations/{celltype}'\n",
    "graph_json_path = './visualization/data/'\n",
    "\n",
    "start_index = 5000\n",
    "end_index = 15000\n",
    "time_based = False\n",
    "\n",
    "import json\n",
    "# with open('/Users/ramprakash/development/lab_projects/Rajapakse_lab/projects/HardWiredGenome/graphs/connected_A_matrix.json') as f:\n",
    "# with open(f'/Users/ramprakash/development/lab_projects/Rajapakse_lab/projects/HardWiredGenome/graphs/combination_perturbed_150_all_cells_{start_index}_{end_index}.json') as f:\n",
    "# with open(f'{HWG_BASE_PATH}/graphs/combination_perturbed_150_all_cells_{start_index}_{end_index}.json') as f:\n",
    "# with open(f'./graphs/combination_perturbed_150_all_cells_{start_index}_{end_index}.json') as f:\n",
    "with open(f'./graphs/combination_perturbed_150_all_cells_{6000}_{7000}.json') as f:\n",
    "    data = json.load(f)\n",
    "node_attrs = data['nodes']\n",
    "\n",
    "# perturbed_edges = []\n",
    "# move_list = []\n",
    "\n",
    "# ENSG00000129152 - MYOD1\n",
    "perturbed_edges = ['ENSG00000129152']\n",
    "move_list = ['ENSG00000129152']\n",
    "\n",
    "# mycn, nanog ,pou5f1\n",
    "# perturbed_edges = ['ENSG00000134323', 'ENSG00000111704', 'ENSG00000204531']\n",
    "# move_list = ['ENSG00000134323', 'ENSG00000111704', 'ENSG00000204531']\n",
    "\n",
    "# yamanaka factors\n",
    "# oct4 - POU5F1\n",
    "# POU5F1, SOX2, KLF4, MYC\n",
    "# ['POU5F1',  'KLF4', 'MYC', 'SOX2']\n",
    "# move_list =  ['ENSG00000204531',  'ENSG00000136826', 'ENSG00000136997', 'ENSG00000181449']\n",
    "\n",
    "\n",
    "# base network\n",
    "\n",
    "filtered_edgelist, filtered_nodes, normal_edges, activator_edges, repressor_edges = get_piece_network(current_edgelist, A_matrix_edgelist, start_index, end_index, flat_col_order, [])\n",
    "filtered_graph = construct_graph(filtered_edgelist, filtered_nodes, normal_edges, activator_edges, repressor_edges, perturbed_edges, node_attrs, time_based=time_based, move_list=move_list)\n",
    "\n",
    "nx.write_graphml(filtered_graph, f\"{savepath}/combination_perturbed_base_time_based_{threshold}_{celltype}_start_end_{start_index}_{end_index}.graphml\")\n",
    "\n",
    "export_graph_as_json(filtered_graph, f\"{graph_json_path}/combination_perturbed_base_time_based_{threshold}_{celltype}_start_end_{start_index}_{end_index}.json\")\n",
    "\n",
    "base_fibroblast_network = filtered_graph.copy()\n",
    "\n",
    "\n",
    "\n",
    "# perturbed_edges = []\n",
    "\n",
    "# first order interactions\n",
    "\n",
    "filtered_edgelist, filtered_nodes, normal_edges, activator_edges, repressor_edges = get_piece_network(current_edgelist, A_matrix_edgelist, start_index, end_index, flat_col_order, perturbed_edges)\n",
    "\n",
    "\n",
    "# filtered_graph = construct_graph(filtered_edgelist, filtered_nodes, perturbed_edges, node_attrs)\n",
    "# # nx.write_gexf(filtered_graph, f\"{savepath}/combination_connected_{threshold}_{celltype}_start_end_{start_index}_{end_index}.gexf\")\n",
    "# nx.write_graphml(filtered_graph, f\"{savepath}/combination_perturbed_first_order_{threshold}_{celltype}_start_end_{start_index}_{end_index}.graphml\")\n",
    "\n",
    "filtered_graph = construct_graph(filtered_edgelist, filtered_nodes, normal_edges, activator_edges, repressor_edges, perturbed_edges, node_attrs, time_based=time_based, move_list=move_list)\n",
    "# nx.write_gexf(filtered_graph, f\"{savepath}/combination_connected_{threshold}_{celltype}_start_end_{start_index}_{end_index}.gexf\")\n",
    "nx.write_graphml(filtered_graph, f\"{savepath}/combination_perturbed_first_order_time_based_{threshold}_{celltype}_start_end_{start_index}_{end_index}.graphml\")\n",
    "\n",
    "export_graph_as_json(filtered_graph, f\"{graph_json_path}/combination_perturbed_first_order_time_based_{threshold}_{celltype}_start_end_{start_index}_{end_index}.json\")\n",
    "first_order_fibroblast_myod_network = filtered_graph.copy()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# second order interactions\n",
    "\n",
    "exists = (filtered_edgelist['source'].isin(perturbed_edges) | (filtered_edgelist['target'].isin(perturbed_edges)))\n",
    "\n",
    "matching_rows = filtered_edgelist[exists]\n",
    "perturbed_edges = list(pd.unique(matching_rows[['source', 'target']].values.ravel()))\n",
    "print(len(perturbed_edges), perturbed_edges)\n",
    "\n",
    "filtered_edgelist, filtered_nodes, normal_edges, activator_edges, repressor_edges = get_piece_network(current_edgelist, A_matrix_edgelist, start_index, end_index, flat_col_order, perturbed_edges)\n",
    "\n",
    "# filtered_graph = construct_graph(filtered_edgelist, filtered_nodes, perturbed_edges, node_attrs)\n",
    "# nx.write_graphml(filtered_graph, f\"{savepath}/combination_perturbed_second_order_{threshold}_{celltype}_start_end_{start_index}_{end_index}.graphml\")\n",
    "\n",
    "filtered_graph = construct_graph(filtered_edgelist, filtered_nodes, normal_edges, activator_edges, repressor_edges, perturbed_edges, node_attrs, time_based=time_based, move_list=move_list)\n",
    "nx.write_graphml(filtered_graph, f\"{savepath}/combination_perturbed_second_order_time_based_{threshold}_{celltype}_start_end_{start_index}_{end_index}.graphml\")\n",
    "\n",
    "export_graph_as_json(filtered_graph, f\"{graph_json_path}/combination_perturbed_second_order_time_based_{threshold}_{celltype}_start_end_{start_index}_{end_index}.json\")\n",
    "\n",
    "second_order_fibroblast_myod_network = filtered_graph.copy()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#third order interactions\n",
    "exists = (filtered_edgelist['source'].isin(perturbed_edges) | (filtered_edgelist['target'].isin(perturbed_edges)))\n",
    "\n",
    "matching_rows = filtered_edgelist[exists]\n",
    "perturbed_edges = list(pd.unique(matching_rows[['source', 'target']].values.ravel()))\n",
    "print(len(perturbed_edges), perturbed_edges)\n",
    "\n",
    "filtered_edgelist, filtered_nodes, normal_edges, activator_edges, repressor_edges = get_piece_network(current_edgelist, A_matrix_edgelist, start_index, end_index, flat_col_order, perturbed_edges)\n",
    "\n",
    "# filtered_graph = construct_graph(filtered_edgelist, filtered_nodes, perturbed_edges, node_attrs)\n",
    "# nx.write_graphml(filtered_graph, f\"{savepath}/combination_perturbed_third_order_{threshold}_{celltype}_start_end_{start_index}_{end_index}.graphml\")\n",
    "\n",
    "filtered_graph = construct_graph(filtered_edgelist, filtered_nodes, normal_edges, activator_edges, repressor_edges, perturbed_edges, node_attrs, time_based=time_based, move_list=move_list)\n",
    "nx.write_graphml(filtered_graph, f\"{savepath}/combination_perturbed_third_order_time_based_{threshold}_{celltype}_start_end_{start_index}_{end_index}.graphml\")\n",
    "\n",
    "export_graph_as_json(filtered_graph, f\"{graph_json_path}/combination_perturbed_third_order_time_based_{threshold}_{celltype}_start_end_{start_index}_{end_index}.json\")\n",
    "\n",
    "third_order_fibroblast_myod_network = filtered_graph.copy()\n",
    "# import pandas as pd\n",
    "\n",
    "# node_data = [\n",
    "#     {\"Id\": node, \"x\": data.get(\"x\", 0), \"y\": data.get(\"y\", 0)}\n",
    "#     for node, data in filtered_graph.nodes(data=True)\n",
    "# ]\n",
    "\n",
    "# pd.DataFrame(node_data).to_csv(f\"{savepath}/combination_perturbed_node_info_{threshold}_{celltype}_start_end_{start_index}_{end_index}.csv\", index=False)\n",
    "\n",
    "\n",
    "# edgelist_df = nx.to_pandas_edgelist(filtered_graph)\n",
    "\n",
    "# edgelist_df.to_csv(f\"{savepath}/combination_connected_{threshold}_{celltype}_start_end_{start_index}_{end_index}.csv\", index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be9fc69a-ef74-4145-b390-3ea27a4f14c9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83c27f7f-0250-4dcf-83e6-9b5a6fbbaa2d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "be43225a-8e1d-4677-ae8c-61d9ebcaeb76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000205609</th>\n",
       "      <th>ENSG00000139579</th>\n",
       "      <th>ENSG00000009307</th>\n",
       "      <th>ENSG00000175329</th>\n",
       "      <th>ENSG00000131910</th>\n",
       "      <th>ENSG00000178028</th>\n",
       "      <th>ENSG00000198538</th>\n",
       "      <th>ENSG00000141052</th>\n",
       "      <th>ENSG00000168724</th>\n",
       "      <th>ENSG00000130254</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000186812</th>\n",
       "      <th>ENSG00000143947</th>\n",
       "      <th>ENSG00000123636</th>\n",
       "      <th>ENSG00000186103</th>\n",
       "      <th>ENSG00000167232</th>\n",
       "      <th>ENSG00000018408</th>\n",
       "      <th>ENSG00000099326</th>\n",
       "      <th>ENSG00000140262</th>\n",
       "      <th>ENSG00000227124</th>\n",
       "      <th>ENSG00000132773</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000112312</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000182968</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000007264</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000113595</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000170004</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000206106</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000206107</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000231068</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243627</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000206140</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18581 rows × 2624 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ENSG00000205609  ENSG00000139579  ENSG00000009307  \\\n",
       "ENSG00000112312                0                0                0   \n",
       "ENSG00000182968                0                0                0   \n",
       "ENSG00000007264                0                0                0   \n",
       "ENSG00000113595                0                0                0   \n",
       "ENSG00000170004                0                0                0   \n",
       "...                          ...              ...              ...   \n",
       "ENSG00000206106                0                0                0   \n",
       "ENSG00000206107                0                0                0   \n",
       "ENSG00000231068                0                0                0   \n",
       "ENSG00000243627                0                0                0   \n",
       "ENSG00000206140                0                0                0   \n",
       "\n",
       "                 ENSG00000175329  ENSG00000131910  ENSG00000178028  \\\n",
       "ENSG00000112312                0                0                0   \n",
       "ENSG00000182968                0                0                0   \n",
       "ENSG00000007264                0                0                0   \n",
       "ENSG00000113595                0                0                0   \n",
       "ENSG00000170004                0                0                0   \n",
       "...                          ...              ...              ...   \n",
       "ENSG00000206106                0                0                0   \n",
       "ENSG00000206107                0                0                0   \n",
       "ENSG00000231068                0                0                0   \n",
       "ENSG00000243627                0                0                0   \n",
       "ENSG00000206140                0                0                0   \n",
       "\n",
       "                 ENSG00000198538  ENSG00000141052  ENSG00000168724  \\\n",
       "ENSG00000112312                0                0                0   \n",
       "ENSG00000182968                0                0                0   \n",
       "ENSG00000007264                0                0                0   \n",
       "ENSG00000113595                0                0                0   \n",
       "ENSG00000170004                0                0                0   \n",
       "...                          ...              ...              ...   \n",
       "ENSG00000206106                0                0                0   \n",
       "ENSG00000206107                0                0                0   \n",
       "ENSG00000231068                0                0                0   \n",
       "ENSG00000243627                0                0                0   \n",
       "ENSG00000206140                0                0                0   \n",
       "\n",
       "                 ENSG00000130254  ...  ENSG00000186812  ENSG00000143947  \\\n",
       "ENSG00000112312                0  ...                0                0   \n",
       "ENSG00000182968                0  ...                0                0   \n",
       "ENSG00000007264                0  ...                0                0   \n",
       "ENSG00000113595                0  ...                0                0   \n",
       "ENSG00000170004                0  ...                0                0   \n",
       "...                          ...  ...              ...              ...   \n",
       "ENSG00000206106                0  ...                0                0   \n",
       "ENSG00000206107                0  ...                0                0   \n",
       "ENSG00000231068                0  ...                0                0   \n",
       "ENSG00000243627                0  ...                0                0   \n",
       "ENSG00000206140                0  ...                0                0   \n",
       "\n",
       "                 ENSG00000123636  ENSG00000186103  ENSG00000167232  \\\n",
       "ENSG00000112312                0                0                0   \n",
       "ENSG00000182968                0                0                0   \n",
       "ENSG00000007264                0                0                0   \n",
       "ENSG00000113595                0                0                0   \n",
       "ENSG00000170004                0                0                0   \n",
       "...                          ...              ...              ...   \n",
       "ENSG00000206106                0                0                0   \n",
       "ENSG00000206107                0                0                0   \n",
       "ENSG00000231068                0                0                0   \n",
       "ENSG00000243627                0                0                0   \n",
       "ENSG00000206140                0                0                0   \n",
       "\n",
       "                 ENSG00000018408  ENSG00000099326  ENSG00000140262  \\\n",
       "ENSG00000112312                0                0                0   \n",
       "ENSG00000182968                0                0                0   \n",
       "ENSG00000007264                0                0                0   \n",
       "ENSG00000113595                0                0                0   \n",
       "ENSG00000170004                0                0                0   \n",
       "...                          ...              ...              ...   \n",
       "ENSG00000206106                0                0                0   \n",
       "ENSG00000206107                0                0                0   \n",
       "ENSG00000231068                0                0                0   \n",
       "ENSG00000243627                0                0                0   \n",
       "ENSG00000206140                0                0                0   \n",
       "\n",
       "                 ENSG00000227124  ENSG00000132773  \n",
       "ENSG00000112312                0                0  \n",
       "ENSG00000182968                0                0  \n",
       "ENSG00000007264                0                0  \n",
       "ENSG00000113595                0                0  \n",
       "ENSG00000170004                0                0  \n",
       "...                          ...              ...  \n",
       "ENSG00000206106                0                0  \n",
       "ENSG00000206107                0                0  \n",
       "ENSG00000231068                0                0  \n",
       "ENSG00000243627                0                0  \n",
       "ENSG00000206140                0                0  \n",
       "\n",
       "[18581 rows x 2624 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from constants import DATA_DIRECTORY \n",
    "\n",
    "BP_matrix = pd.read_csv(f'{DATA_DIRECTORY}/HWG/BP_Matrix.csv', index_col=0)\n",
    "\n",
    "\n",
    "BP_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "88807d74-1cdf-41b1-a391-fd25a576fbe2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Activator counts:  2563\n",
      "Repressor counts:  1646\n",
      "Unknown counts:  48752335\n"
     ]
    }
   ],
   "source": [
    "count = (BP_matrix == 1).sum().sum()\n",
    "print(\"Activator counts: \", count)\n",
    "\n",
    "count = (BP_matrix == -1).sum().sum()\n",
    "print(\"Repressor counts: \", count)\n",
    "\n",
    "count = (BP_matrix == 0).sum().sum()\n",
    "print(\"Unknown counts: \", count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fe7fc75-0791-48bc-a7aa-76cc551d9c7b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1ae49d2-77e6-4a02-91b6-5fed0d3a8c6c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f740020-fc30-45fc-b613-b04eff8ecf47",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90a15bc7-e4c8-4bc9-a3fb-beb4d24e396f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# \n",
    "celltype = 'fibroblast'\n",
    "# celltype = 'myotube'\n",
    "# celltype = 'esc'\n",
    "# celltype = 'all_cells'\n",
    "# current_edgelist = A_matrix_edgelist\n",
    "current_edgelist = fibroblast_filtered_edgelist\n",
    "# current_edgelist = myotube_filtered_edgelist\n",
    "# current_edgelist = esc_filtered_edgelist\n",
    "\n",
    "savepath = f'{HWG_BASE_PATH}/combinations/{celltype}'\n",
    "graph_json_path = './visualization/data/'\n",
    "\n",
    "start_index = 6000\n",
    "end_index = 7000\n",
    "time_based = False\n",
    "\n",
    "import json\n",
    "# with open('/Users/ramprakash/development/lab_projects/Rajapakse_lab/projects/HardWiredGenome/graphs/connected_A_matrix.json') as f:\n",
    "# with open(f'/Users/ramprakash/development/lab_projects/Rajapakse_lab/projects/HardWiredGenome/graphs/combination_perturbed_150_all_cells_{start_index}_{end_index}.json') as f:\n",
    "# with open(f'{HWG_BASE_PATH}/graphs/combination_perturbed_150_all_cells_{start_index}_{end_index}.json') as f:\n",
    "with open(f'./graphs/combination_perturbed_150_all_cells_{start_index}_{end_index}.json') as f:\n",
    "# with open(f'./graphs/combination_perturbed_150_all_cells_{6000}_{7000}.json') as f:\n",
    "    data = json.load(f)\n",
    "node_attrs = data['nodes']\n",
    "\n",
    "\n",
    "\n",
    "perturbed_edges = []\n",
    "move_list = []\n",
    "\n",
    "\n",
    "# base network\n",
    "\n",
    "filtered_edgelist, filtered_nodes, normal_edges, activator_edges, repressor_edges = get_piece_network(current_edgelist, A_matrix_edgelist, start_index, end_index, flat_col_order, [])\n",
    "filtered_graph = construct_graph(filtered_edgelist, filtered_nodes, normal_edges, activator_edges, repressor_edges, perturbed_edges, node_attrs, time_based=time_based, move_list=move_list)\n",
    "\n",
    "nx.write_graphml(filtered_graph, f\"{savepath}/combination_perturbed_base_time_based_{threshold}_{celltype}_start_end_{start_index}_{end_index}.graphml\")\n",
    "\n",
    "export_graph_as_json(filtered_graph, f\"{graph_json_path}/combination_perturbed_base_time_based_{threshold}_{celltype}_start_end_{start_index}_{end_index}.json\")\n",
    "\n",
    "base_myotube_network = filtered_graph.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47cd9574-b314-401c-8f67-e8ae75d3d9e8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "cacfa6c9-0595-4bc8-879c-221715d248a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [source, target, weight]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [source, target, weight]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "s_edges = fibroblast_filtered_edgelist[fibroblast_filtered_edgelist['source'] == 'ENSG00000129152']\n",
    "t_edges = fibroblast_filtered_edgelist[fibroblast_filtered_edgelist['target'] == 'ENSG00000129152']\n",
    "\n",
    "print(s_edges)\n",
    "print(t_edges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "5888c3b3-2125-448c-806d-aaa953a82486",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [source, target, weight]\n",
      "Index: []\n",
      "Empty DataFrame\n",
      "Columns: [source, target, weight]\n",
      "Index: []\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Columns must be same length as key",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mValueError\u001b[39m                                Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[117]\u001b[39m\u001b[32m, line 33\u001b[39m\n\u001b[32m     29\u001b[39m perturbed_edgelist[\u001b[33m'\u001b[39m\u001b[33medge\u001b[39m\u001b[33m'\u001b[39m] = perturbed_edgelist.apply(\u001b[38;5;28;01mlambda\u001b[39;00m row: \u001b[38;5;28mtuple\u001b[39m(\u001b[38;5;28msorted\u001b[39m([row[\u001b[33m'\u001b[39m\u001b[33msource\u001b[39m\u001b[33m'\u001b[39m], row[\u001b[33m'\u001b[39m\u001b[33mtarget\u001b[39m\u001b[33m'\u001b[39m]])), axis=\u001b[32m1\u001b[39m)\n\u001b[32m     31\u001b[39m df_unique = perturbed_edgelist.drop_duplicates(\u001b[33m'\u001b[39m\u001b[33medge\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m---> \u001b[39m\u001b[32m33\u001b[39m \u001b[43mdf_unique\u001b[49m\u001b[43m[\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43msource\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mtarget\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m = pd.DataFrame(df_unique[\u001b[33m'\u001b[39m\u001b[33medge\u001b[39m\u001b[33m'\u001b[39m].tolist(), index=df_unique.index)\n\u001b[32m     34\u001b[39m df = df_unique[[\u001b[33m'\u001b[39m\u001b[33msource\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mtarget\u001b[39m\u001b[33m'\u001b[39m]]\n\u001b[32m     36\u001b[39m mask = df[\u001b[33m'\u001b[39m\u001b[33mtarget\u001b[39m\u001b[33m'\u001b[39m].isin(activator_edges) & ~df[\u001b[33m'\u001b[39m\u001b[33msource\u001b[39m\u001b[33m'\u001b[39m].isin(activator_edges)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.local/lib/python3.11/site-packages/pandas/core/frame.py:4299\u001b[39m, in \u001b[36mDataFrame.__setitem__\u001b[39m\u001b[34m(self, key, value)\u001b[39m\n\u001b[32m   4297\u001b[39m     \u001b[38;5;28mself\u001b[39m._setitem_frame(key, value)\n\u001b[32m   4298\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, (Series, np.ndarray, \u001b[38;5;28mlist\u001b[39m, Index)):\n\u001b[32m-> \u001b[39m\u001b[32m4299\u001b[39m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_setitem_array\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   4300\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(value, DataFrame):\n\u001b[32m   4301\u001b[39m     \u001b[38;5;28mself\u001b[39m._set_item_frame_value(key, value)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.local/lib/python3.11/site-packages/pandas/core/frame.py:4341\u001b[39m, in \u001b[36mDataFrame._setitem_array\u001b[39m\u001b[34m(self, key, value)\u001b[39m\n\u001b[32m   4336\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m   4337\u001b[39m     \u001b[38;5;66;03m# Note: unlike self.iloc[:, indexer] = value, this will\u001b[39;00m\n\u001b[32m   4338\u001b[39m     \u001b[38;5;66;03m#  never try to overwrite values inplace\u001b[39;00m\n\u001b[32m   4340\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(value, DataFrame):\n\u001b[32m-> \u001b[39m\u001b[32m4341\u001b[39m         \u001b[43mcheck_key_length\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   4342\u001b[39m         \u001b[38;5;28;01mfor\u001b[39;00m k1, k2 \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(key, value.columns):\n\u001b[32m   4343\u001b[39m             \u001b[38;5;28mself\u001b[39m[k1] = value[k2]\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.local/lib/python3.11/site-packages/pandas/core/indexers/utils.py:390\u001b[39m, in \u001b[36mcheck_key_length\u001b[39m\u001b[34m(columns, key, value)\u001b[39m\n\u001b[32m    388\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m columns.is_unique:\n\u001b[32m    389\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(value.columns) != \u001b[38;5;28mlen\u001b[39m(key):\n\u001b[32m--> \u001b[39m\u001b[32m390\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[33m\"\u001b[39m\u001b[33mColumns must be same length as key\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m    391\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    392\u001b[39m     \u001b[38;5;66;03m# Missing keys in columns are represented as -1\u001b[39;00m\n\u001b[32m    393\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(columns.get_indexer_non_unique(key)[\u001b[32m0\u001b[39m]) != \u001b[38;5;28mlen\u001b[39m(value.columns):\n",
      "\u001b[31mValueError\u001b[39m: Columns must be same length as key"
     ]
    }
   ],
   "source": [
    "threshold = 0\n",
    "# path = HARD_WIRED_GENOME_A_MATRIX_PATH.strip('.csv') + f\"_{threshold}.csv\"\n",
    "\n",
    "# path =  HARD_WIRED_GENOME_A_EDGELIST.strip('.csv') + f\"_{threshold}.csv\"\n",
    "\n",
    "# A_matrix_edgelist = pd.read_csv(path, index_col=0)\n",
    "\n",
    "# full_edgelist = A_matrix_edgelist\n",
    "# full_edgelist = fibroblast_filtered_edgelist\n",
    "\n",
    "\n",
    "\n",
    "activator_edges = ['ENSG00000129152']\n",
    "\n",
    "\n",
    "ts_perturbed_edgelist = full_edgelist[\n",
    "        full_edgelist['source'].isin(activator_edges) & full_edgelist['target'].isin(flat_col_order)\n",
    "]\n",
    "print(ts_perturbed_edgelist)\n",
    "\n",
    "st_perturbed_edgelist = full_edgelist[\n",
    "        full_edgelist['source'].isin(flat_col_order) & full_edgelist['target'].isin(activator_edges)\n",
    "]\n",
    "print(st_perturbed_edgelist)\n",
    "\n",
    "perturbed_edgelist = pd.concat([ts_perturbed_edgelist, st_perturbed_edgelist], ignore_index=True)\n",
    "\n",
    "\n",
    "perturbed_edgelist['edge'] = perturbed_edgelist.apply(lambda row: tuple(sorted([row['source'], row['target']])), axis=1)\n",
    "\n",
    "df_unique = perturbed_edgelist.drop_duplicates('edge')\n",
    "\n",
    "df_unique[['source', 'target']] = pd.DataFrame(df_unique['edge'].tolist(), index=df_unique.index)\n",
    "df = df_unique[['source', 'target']]\n",
    "\n",
    "mask = df['target'].isin(activator_edges) & ~df['source'].isin(activator_edges)\n",
    "\n",
    "df.loc[mask, ['source', 'target']] = df.loc[mask, ['target', 'source']].values\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "70189f3c-34e4-4f61-a3e8-7d58bf994c55",
   "metadata": {},
   "outputs": [],
   "source": [
    "myod_targets = list(set(df['target']))\n",
    "original_potential_targets = len(myod_targets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "f7a688c3-e35b-4625-af06-ccf73dd8b898",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Myod Targets (all Cell Types) : 1612\n",
      "Existing Activated Genes (Fibroblast) : 325\n",
      "Newly MYOD activated Genes (Fibroblast) : 1287\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "c5798ff7-fd00-42ab-b777-5be2d3d05ec5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processed 1639 Transcription Factors\n",
      "['DUX1_HUMAN', 'DUX3_HUMAN', 'ENSG00000001167', 'ENSG00000004848', 'ENSG00000005073', 'ENSG00000005102', 'ENSG00000005513', 'ENSG00000005801', 'ENSG00000005889', 'ENSG00000006047']\n"
     ]
    }
   ],
   "source": [
    "from constants import HUMAN_TF_IDLIST_PATH\n",
    "\n",
    "def get_TF_set():\n",
    "    tflist = []\n",
    "    with open(HUMAN_TF_IDLIST_PATH, 'r') as f:\n",
    "        for line in f.readlines():\n",
    "            tflist.append(line.strip())\n",
    "            \n",
    "    print(f'processed {len(tflist)} Transcription Factors')\n",
    "    return tflist\n",
    "\n",
    "tfset = get_TF_set()\n",
    "print(tfset[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "bbcdcbbf-8007-4dcb-8982-7c0279c54b34",
   "metadata": {},
   "outputs": [],
   "source": [
    "c_matrix = C_matrix\n",
    "\n",
    "master_regulators = c_matrix[c_matrix['source'] == c_matrix['target']]\n",
    "\n",
    "master_regulator_list = list(master_regulators['source'])\n",
    "\n",
    "transcription_factor_matx = c_matrix[c_matrix['source'] != c_matrix['target']]\n",
    "\n",
    "transcription_factor_list = pd.unique(transcription_factor_matx[['source', 'target']].values.ravel())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "6a908a66-89f2-4875-889e-25ed2cd5bdf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "master_regulator_set = set(master_regulator_list)\n",
    "tf_set = set(transcription_factor_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "ea618319-ffd4-4c13-9636-b66e57bc9cba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Myod Targets (all Cell Types) : 1612\n",
      "Existing Activated Genes (Fibroblast) : 325\n",
      "Newly MYOD activated Genes (Fibroblast) : 1287\n",
      "--------------------------\n",
      "Master Regulators within targets : 52\n",
      "Transcription Factors within targets : 414\n",
      "Normal Genes within targets : 1146\n"
     ]
    }
   ],
   "source": [
    "fibset = set(fibroblast_expressed_genes)\n",
    "targetset = set(myod_targets)\n",
    "\n",
    "\n",
    "print(f'Total Myod Targets (all Cell Types) : {len(targetset)}')\n",
    "# print(f'Total Expressed Genes in fibroblast : {len(fibset)}')\n",
    "print(f'Existing Activated Genes (Fibroblast) : {len(targetset.intersection(fibset))}')\n",
    "print(f'Newly MYOD activated Genes (Fibroblast) : {len(targetset - targetset.intersection(fibset))}')\n",
    "\n",
    "print('--------------------------')\n",
    "print(f'Master Regulators within targets : {len(targetset.intersection(master_regulator_set))}')\n",
    "print(f'Transcription Factors within targets : {len(targetset.intersection(tf_set) - master_regulator_set)}')\n",
    "print(f'Normal Genes within targets : {len(targetset - tf_set)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "db34ba93-0205-4a59-bd15-9718fec3b025",
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify_node(row):\n",
    "    if row[\"Gene ID\"] in master_regulator_set:\n",
    "        return \"g1\"\n",
    "    elif row[\"Gene ID\"] in tf_set:\n",
    "        return \"g2\"\n",
    "    else:\n",
    "        return \"g3\"\n",
    "\n",
    "def add_name(row):\n",
    "    return gene_id_name_map.get(row[\"Gene ID\"], \"NA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "c00945b3-8a53-424f-8670-7bcb914b70c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "fib_G = nx.from_pandas_edgelist(fibroblast_filtered_edgelist, 'source', 'target', edge_attr='weight', create_using=nx.Graph())\n",
    "\n",
    "fib_G.add_nodes_from(list(targetset))\n",
    "\n",
    "\n",
    "degree_centrality = nx.degree_centrality(fib_G)\n",
    "eigen_centrality = nx.eigenvector_centrality(fib_G, max_iter=1000)\n",
    "\n",
    "df_deg = pd.DataFrame(degree_centrality.items(), columns=[\"Gene ID\", \"Degree Centrality\"])\n",
    "df_eig = pd.DataFrame(eigen_centrality.items(), columns=[\"Gene ID\", \"Eigenvector Centrality\"])\n",
    "\n",
    "df_all = pd.merge(df_deg, df_eig, on=\"Gene ID\")\n",
    "df_all = df_all.round(4)\n",
    "\n",
    "df_all[\"Gene Type\"] = df_all.apply(classify_node, axis=1)\n",
    "df_all[\"Gene Name\"] = df_all.apply(add_name, axis=1)\n",
    "\n",
    "\n",
    "df_targets = df_all[df_all[\"Gene ID\"].isin(targetset)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "bcf3dbfb-ed09-41ea-84d9-38d48c1b4434",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_targets.to_csv('./data/myod_first_order_interaction_targets.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "f8e19038-9fd8-4770-b47a-bc1e22537065",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Degree Centrality</th>\n",
       "      <th>Eigenvector Centrality</th>\n",
       "      <th>Gene Type</th>\n",
       "      <th>Gene Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000063978</td>\n",
       "      <td>0.1605</td>\n",
       "      <td>0.0259</td>\n",
       "      <td>g3</td>\n",
       "      <td>RNF4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000142937</td>\n",
       "      <td>0.1766</td>\n",
       "      <td>0.0397</td>\n",
       "      <td>g3</td>\n",
       "      <td>RPS8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000198721</td>\n",
       "      <td>0.0940</td>\n",
       "      <td>0.0178</td>\n",
       "      <td>g3</td>\n",
       "      <td>ECI2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000117118</td>\n",
       "      <td>0.1875</td>\n",
       "      <td>0.0358</td>\n",
       "      <td>g3</td>\n",
       "      <td>SDHB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000183722</td>\n",
       "      <td>0.0275</td>\n",
       "      <td>0.0023</td>\n",
       "      <td>g3</td>\n",
       "      <td>LHFPL6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3666</th>\n",
       "      <td>ENSG00000177606</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>g1</td>\n",
       "      <td>JUN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3667</th>\n",
       "      <td>ENSG00000135457</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>g2</td>\n",
       "      <td>TFCP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3668</th>\n",
       "      <td>ENSG00000187642</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>g3</td>\n",
       "      <td>PERM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3669</th>\n",
       "      <td>ENSG00000141086</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>g3</td>\n",
       "      <td>CTRL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3670</th>\n",
       "      <td>ENSG00000092607</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>g1</td>\n",
       "      <td>TBX15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3671 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Gene ID  Degree Centrality  Eigenvector Centrality Gene Type  \\\n",
       "0     ENSG00000063978             0.1605                  0.0259        g3   \n",
       "1     ENSG00000142937             0.1766                  0.0397        g3   \n",
       "2     ENSG00000198721             0.0940                  0.0178        g3   \n",
       "3     ENSG00000117118             0.1875                  0.0358        g3   \n",
       "4     ENSG00000183722             0.0275                  0.0023        g3   \n",
       "...               ...                ...                     ...       ...   \n",
       "3666  ENSG00000177606             0.0000                  0.0000        g1   \n",
       "3667  ENSG00000135457             0.0000                  0.0000        g2   \n",
       "3668  ENSG00000187642             0.0000                  0.0000        g3   \n",
       "3669  ENSG00000141086             0.0000                  0.0000        g3   \n",
       "3670  ENSG00000092607             0.0000                  0.0000        g1   \n",
       "\n",
       "     Gene Name  \n",
       "0         RNF4  \n",
       "1         RPS8  \n",
       "2         ECI2  \n",
       "3         SDHB  \n",
       "4       LHFPL6  \n",
       "...        ...  \n",
       "3666       JUN  \n",
       "3667     TFCP2  \n",
       "3668     PERM1  \n",
       "3669      CTRL  \n",
       "3670     TBX15  \n",
       "\n",
       "[3671 rows x 5 columns]"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "012fb1ad-e188-4c3d-b321-e0bf31fa60cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Degree Centrality Eigenvector Centrality\n",
      "GAPDH                GAPDH                    \n",
      "ACTB                 ACTB                     \n",
      "AKT1                 HSP90AA1                 \n",
      "HSP90AA1             HSP90AB1                 \n",
      "HSP90AB1             HSPA4                    \n",
      "CTNNB1               HSPA8                    \n",
      "HSPA4                EEF2                     \n",
      "HSPA8                HSPA5                    \n",
      "HSPA5                AKT1                     \n",
      "RHOA                 HSPA9                    \n"
     ]
    }
   ],
   "source": [
    "print(\"Degree Centrality\", \"Eigenvector Centrality\")\n",
    "cnt = 10\n",
    "top_degsh =  list(df_all.sort_values(by=\"Degree Centrality\", ascending=False)[:cnt]['Gene Name'])\n",
    "top_egsh = list(df_all.sort_values(by=\"Eigenvector Centrality\", ascending=False)[:cnt]['Gene Name'])\n",
    "for i in range(cnt):\n",
    "    # print(top_degsh[i], top_egsh[i])\n",
    "    print(f\"{top_degsh[i]:<20} {top_egsh[i]:<25}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "117eb031-c540-49a7-b0b7-414d915dda14",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GAPDH\n",
      "ACTB\n",
      "HSP90AA1\n",
      "HSP90AB1\n",
      "HSPA4\n",
      "HSPA8\n",
      "EEF2\n",
      "HSPA5\n",
      "AKT1\n",
      "HSPA9\n"
     ]
    }
   ],
   "source": [
    "for i in :\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "4135b39e-146c-4f60-aaba-a8fe0e5f98f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2638"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(fibroblast_expressed_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "dd858c53-8afc-4ede-9329-2e84ca5ffe3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "132"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(master_regulator_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de1a26c2-026c-41cc-935e-b0ea44fcd068",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HardWiredGenome",
   "language": "python",
   "name": "hardwiredgenome"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
