{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fb6e0e4f-6aed-4a1e-abd1-edd0a6cc2225",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import numpy as np\n",
    "\n",
    "from anndata import AnnData\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "aecde193-b2cc-4a9e-9098-582fb54d47f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracted 43945 mappings from 73467 ensembl lines\n",
      "Extracted 38606 mappings from gene id gtf file\n",
      "Total Mappings Extracted 59979\n"
     ]
    }
   ],
   "source": [
    "from mapping import generate_gene_id_name_map \n",
    "\n",
    "gene_id_name_map, gene_name_id_map = generate_gene_id_name_map()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d4d233e-101e-4871-b950-6c593a3b55b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "gtf_file = '/nfs/turbo/umms-indikar/shared/projects/reference_genome/prebuilt/refdata-gex-GRCh38-2024-A/genes/genes.gtf'\n",
    "\n",
    "df = pd.read_csv(gtf_file, sep='\\t', header=None, comment='#')\n",
    "\n",
    "df_back = df.copy()\n",
    "\n",
    "df = df[df[2] == 'gene']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b769f305-98a9-4ff6-8503-ba96b3b17c2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_gtf(dfc):\n",
    "    df = dfc.copy()\n",
    "    df.columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']\n",
    "    \n",
    "    fields = ['gene_id', 'transcript_id', 'exon_number', 'gene_name', 'gene_source', 'gene_biotype', 'transcript_name', 'transcript_source', 'transcript_biotype', 'protein_id', 'exon_id', 'tag']\n",
    "    for field in fields:\n",
    "        df[field] = df['attribute'].apply(lambda x: re.findall(rf'{field} \"([^\"]*)\"', x)[0] if rf'{field} \"' in x else '')\n",
    "    \n",
    "    df.replace('', np.nan, inplace=True)\n",
    "    \n",
    "    df.drop('attribute', axis=1, inplace=True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f48b6374-1a1f-436b-b4cd-74e4f6562a32",
   "metadata": {},
   "outputs": [],
   "source": [
    "formatted_df = format_gtf(df)\n",
    "mapping_dict = dict(zip(formatted_df['gene_id'], formatted_df['gene_name']))\n",
    "\n",
    "formatted_df[['gene_id', 'gene_name']].to_csv('./data/gene_id_name_map_gtf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4c8a3164-065a-4ed4-bfac-0d44e9924631",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "gene_metadata = pd.read_csv('./data/DSET038_37_Combined_GRCh38p14_GRCh37p13.txt')\n",
    "gene_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00d7ec23-faf3-4374-a8d5-16b139f2c3fa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "267fad10-6092-4fb0-8fe0-35f0b2cbf257",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "eb556107-e403-479d-a30e-0758b1d42cd9",
   "metadata": {},
   "source": [
    "### B Matrix Operations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2dc7d426-3f37-404c-bf3b-3f402e623282",
   "metadata": {},
   "outputs": [],
   "source": [
    "B_MATRIX_PATH = '/scratch/indikar_root/indikar1/shared_data/HWG/operations'\n",
    "\n",
    "X = pd.read_csv(f'{B_MATRIX_PATH}/DSET051_B_Matrix_TFsONLY_TFTGDb_TFLink_ChEA3_X.csv', header=None, index_col=None)\n",
    "X_arr = X.to_numpy()\n",
    "obs = pd.read_csv(f'{B_MATRIX_PATH}/DSET051_B_Matrix_TFsONLY_TFTGDb_TFLink_ChEA3_obs.csv')\n",
    "var = pd.read_csv(f'{B_MATRIX_PATH}/DSET051_B_Matrix_TFsONLY_TFTGDb_TFLink_ChEA3_var.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "de6a8c8f-3fad-4a17-81a8-be4331ea5cfa",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/nuvi/.local/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/home/nuvi/.local/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    }
   ],
   "source": [
    "adata = AnnData(X=X_arr, obs=obs, var=var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "db3da8ed-30e4-4dfc-9fb3-d4edbb8bd4d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TFStableIDVersion</th>\n",
       "      <th>TFStableID</th>\n",
       "      <th>TFName</th>\n",
       "      <th>TFClass</th>\n",
       "      <th>IsTF</th>\n",
       "      <th>GeneVersion</th>\n",
       "      <th>GeneDescription</th>\n",
       "      <th>Entrez_Description</th>\n",
       "      <th>GeneType</th>\n",
       "      <th>ChromosomescaffoldName</th>\n",
       "      <th>...</th>\n",
       "      <th>Motif_Status</th>\n",
       "      <th>ORFeome_Activator</th>\n",
       "      <th>DelRosso_Activators</th>\n",
       "      <th>DelRosso_Repressors</th>\n",
       "      <th>DelRosso_Bifunctional</th>\n",
       "      <th>TRRUST_Activators</th>\n",
       "      <th>TRRUST_Repressors</th>\n",
       "      <th>TRRUST_Conflicting</th>\n",
       "      <th>Sums</th>\n",
       "      <th>Target_Gene_Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000001167.15</td>\n",
       "      <td>ENSG00000001167</td>\n",
       "      <td>NFYA</td>\n",
       "      <td>Activator</td>\n",
       "      <td>Yes</td>\n",
       "      <td>15</td>\n",
       "      <td>DNA binding; regulation of DNA-templated trans...</td>\n",
       "      <td>nuclear transcription factor Y subunit alpha</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>In vivo/Misc source</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>15524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000004848.8</td>\n",
       "      <td>ENSG00000004848</td>\n",
       "      <td>ARX</td>\n",
       "      <td>Repressor</td>\n",
       "      <td>Yes</td>\n",
       "      <td>8</td>\n",
       "      <td>DNA binding; regulation of DNA-templated trans...</td>\n",
       "      <td>aristaless related homeobox</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>X</td>\n",
       "      <td>...</td>\n",
       "      <td>High-throughput in vitro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1</td>\n",
       "      <td>402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000005073.6</td>\n",
       "      <td>ENSG00000005073</td>\n",
       "      <td>HOXA11</td>\n",
       "      <td>Repressor</td>\n",
       "      <td>Yes</td>\n",
       "      <td>6</td>\n",
       "      <td>DNA binding; regulation of DNA-templated trans...</td>\n",
       "      <td>homeobox A11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>High-throughput in vitro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2</td>\n",
       "      <td>381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000005102.14</td>\n",
       "      <td>ENSG00000005102</td>\n",
       "      <td>MEOX1</td>\n",
       "      <td>Conflicted</td>\n",
       "      <td>Yes</td>\n",
       "      <td>14</td>\n",
       "      <td>DNA binding; regulation of DNA-templated trans...</td>\n",
       "      <td>mesenchyme homeobox 1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>High-throughput in vitro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000005513.11</td>\n",
       "      <td>ENSG00000005513</td>\n",
       "      <td>SOX8</td>\n",
       "      <td>Conflicted</td>\n",
       "      <td>Yes</td>\n",
       "      <td>11</td>\n",
       "      <td>organelle; nucleus; DNA binding; reproductive ...</td>\n",
       "      <td>SRY-box transcription factor 8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>High-throughput in vitro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>5308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1603</th>\n",
       "      <td>ENSG00000277258.5</td>\n",
       "      <td>ENSG00000277258</td>\n",
       "      <td>PCGF2</td>\n",
       "      <td>Repressor</td>\n",
       "      <td>Yes</td>\n",
       "      <td>5</td>\n",
       "      <td>organelle; nucleus; nucleoplasm; DNA binding; ...</td>\n",
       "      <td>polycomb group ring finger 2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>No motif</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2</td>\n",
       "      <td>3022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1604</th>\n",
       "      <td>ENSG00000277462.2</td>\n",
       "      <td>ENSG00000277462</td>\n",
       "      <td>ZNF670</td>\n",
       "      <td>Repressor</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>regulation of DNA-templated transcription; org...</td>\n",
       "      <td>zinc finger protein 670</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>No motif</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1</td>\n",
       "      <td>458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1605</th>\n",
       "      <td>ENSG00000278129.2</td>\n",
       "      <td>ENSG00000278129</td>\n",
       "      <td>ZNF8</td>\n",
       "      <td>Conflicted</td>\n",
       "      <td>Yes</td>\n",
       "      <td>2</td>\n",
       "      <td>regulation of DNA-templated transcription; org...</td>\n",
       "      <td>zinc finger protein 8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>19</td>\n",
       "      <td>...</td>\n",
       "      <td>High-throughput in vitro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1606</th>\n",
       "      <td>ENSG00000278318.5</td>\n",
       "      <td>ENSG00000278318</td>\n",
       "      <td>ZNF229</td>\n",
       "      <td>Conflicted</td>\n",
       "      <td>Yes</td>\n",
       "      <td>5</td>\n",
       "      <td>regulation of DNA-templated transcription; org...</td>\n",
       "      <td>zinc finger protein 229</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>19</td>\n",
       "      <td>...</td>\n",
       "      <td>No motif</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1607</th>\n",
       "      <td>ENSG00000278570.5</td>\n",
       "      <td>ENSG00000278570</td>\n",
       "      <td>NR2E3</td>\n",
       "      <td>Activator</td>\n",
       "      <td>Yes</td>\n",
       "      <td>5</td>\n",
       "      <td>regulation of DNA-templated transcription; tra...</td>\n",
       "      <td>nuclear receptor subfamily 2 group E member 3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>High-throughput in vitro</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>417</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1608 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       TFStableIDVersion       TFStableID  TFName     TFClass IsTF  \\\n",
       "0     ENSG00000001167.15  ENSG00000001167    NFYA   Activator  Yes   \n",
       "1      ENSG00000004848.8  ENSG00000004848     ARX   Repressor  Yes   \n",
       "2      ENSG00000005073.6  ENSG00000005073  HOXA11   Repressor  Yes   \n",
       "3     ENSG00000005102.14  ENSG00000005102   MEOX1  Conflicted  Yes   \n",
       "4     ENSG00000005513.11  ENSG00000005513    SOX8  Conflicted  Yes   \n",
       "...                  ...              ...     ...         ...  ...   \n",
       "1603   ENSG00000277258.5  ENSG00000277258   PCGF2   Repressor  Yes   \n",
       "1604   ENSG00000277462.2  ENSG00000277462  ZNF670   Repressor  Yes   \n",
       "1605   ENSG00000278129.2  ENSG00000278129    ZNF8  Conflicted  Yes   \n",
       "1606   ENSG00000278318.5  ENSG00000278318  ZNF229  Conflicted  Yes   \n",
       "1607   ENSG00000278570.5  ENSG00000278570   NR2E3   Activator  Yes   \n",
       "\n",
       "      GeneVersion                                    GeneDescription  \\\n",
       "0              15  DNA binding; regulation of DNA-templated trans...   \n",
       "1               8  DNA binding; regulation of DNA-templated trans...   \n",
       "2               6  DNA binding; regulation of DNA-templated trans...   \n",
       "3              14  DNA binding; regulation of DNA-templated trans...   \n",
       "4              11  organelle; nucleus; DNA binding; reproductive ...   \n",
       "...           ...                                                ...   \n",
       "1603            5  organelle; nucleus; nucleoplasm; DNA binding; ...   \n",
       "1604            2  regulation of DNA-templated transcription; org...   \n",
       "1605            2  regulation of DNA-templated transcription; org...   \n",
       "1606            5  regulation of DNA-templated transcription; org...   \n",
       "1607            5  regulation of DNA-templated transcription; tra...   \n",
       "\n",
       "                                 Entrez_Description        GeneType  \\\n",
       "0      nuclear transcription factor Y subunit alpha  protein_coding   \n",
       "1                       aristaless related homeobox  protein_coding   \n",
       "2                                      homeobox A11  protein_coding   \n",
       "3                             mesenchyme homeobox 1  protein_coding   \n",
       "4                    SRY-box transcription factor 8  protein_coding   \n",
       "...                                             ...             ...   \n",
       "1603                   polycomb group ring finger 2  protein_coding   \n",
       "1604                        zinc finger protein 670  protein_coding   \n",
       "1605                          zinc finger protein 8  protein_coding   \n",
       "1606                        zinc finger protein 229  protein_coding   \n",
       "1607  nuclear receptor subfamily 2 group E member 3  protein_coding   \n",
       "\n",
       "     ChromosomescaffoldName  ...              Motif_Status  ORFeome_Activator  \\\n",
       "0                         6  ...       In vivo/Misc source                NaN   \n",
       "1                         X  ...  High-throughput in vitro                NaN   \n",
       "2                         7  ...  High-throughput in vitro                NaN   \n",
       "3                        17  ...  High-throughput in vitro                NaN   \n",
       "4                        16  ...  High-throughput in vitro                NaN   \n",
       "...                     ...  ...                       ...                ...   \n",
       "1603                     17  ...                  No motif                NaN   \n",
       "1604                      1  ...                  No motif                NaN   \n",
       "1605                     19  ...  High-throughput in vitro                NaN   \n",
       "1606                     19  ...                  No motif                NaN   \n",
       "1607                     15  ...  High-throughput in vitro                NaN   \n",
       "\n",
       "      DelRosso_Activators  DelRosso_Repressors DelRosso_Bifunctional  \\\n",
       "0                     NaN                  NaN                   NaN   \n",
       "1                     NaN                 -1.0                   NaN   \n",
       "2                     NaN                 -1.0                   NaN   \n",
       "3                     NaN                  NaN                   NaN   \n",
       "4                     1.0                 -1.0                   NaN   \n",
       "...                   ...                  ...                   ...   \n",
       "1603                  NaN                 -1.0                   NaN   \n",
       "1604                  NaN                 -1.0                   NaN   \n",
       "1605                  NaN                  NaN                   NaN   \n",
       "1606                  NaN                  NaN                   NaN   \n",
       "1607                  NaN                  NaN                   NaN   \n",
       "\n",
       "     TRRUST_Activators TRRUST_Repressors TRRUST_Conflicting  Sums  \\\n",
       "0                  1.0               NaN                NaN     1   \n",
       "1                  NaN               NaN                0.0    -1   \n",
       "2                  NaN              -1.0                NaN    -2   \n",
       "3                  NaN               NaN                NaN     0   \n",
       "4                  NaN               NaN                NaN     0   \n",
       "...                ...               ...                ...   ...   \n",
       "1603               NaN              -1.0                NaN    -2   \n",
       "1604               NaN               NaN                NaN    -1   \n",
       "1605               NaN               NaN                NaN     0   \n",
       "1606               NaN               NaN                NaN     0   \n",
       "1607               1.0               NaN                NaN     1   \n",
       "\n",
       "      Target_Gene_Count  \n",
       "0                 15524  \n",
       "1                   402  \n",
       "2                   381  \n",
       "3                   977  \n",
       "4                  5308  \n",
       "...                 ...  \n",
       "1603               3022  \n",
       "1604                458  \n",
       "1605               1171  \n",
       "1606                456  \n",
       "1607                417  \n",
       "\n",
       "[1608 rows x 27 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "448287fd-9c31-422f-96fc-e417f3ecdd16",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(f'{B_MATRIX_PATH}/DSET051_B_Matrix_TFsONLY_TFTGDb_TFLink_ChEA3.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6106c6a0-228a-477e-9abe-6fe4554751c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "tenk_b_matrix = pd.read_csv('/nfs/turbo/umms-indikar/shared/projects/DGC/data/processed_data/10kbp_up_10kbp_down_B.csv')\n",
    "\n",
    "five_h_b_matrix = pd.read_csv('/nfs/turbo/umms-indikar/shared/projects/DGC/data/processed_data/500bp_up_100bp_down_B.csv')\n",
    "\n",
    "tenk_b_matrix['GeneName'] = tenk_b_matrix['Unnamed: 0']\n",
    "tenk_b_matrix.drop('Unnamed: 0', axis=1, inplace=True)\n",
    "\n",
    "five_h_b_matrix['GeneName'] = five_h_b_matrix['Unnamed: 0']\n",
    "five_h_b_matrix.drop('Unnamed: 0', axis=1, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0762a820-8c9b-4a4e-b9f6-88cd4deba152",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ordering 18581 genes\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "18581"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from mapping import get_sorted_gene_order\n",
    "\n",
    "flat_col_order = get_sorted_gene_order()\n",
    "\n",
    "nodes = flat_col_order\n",
    "nodenames = [gene_id_name_map.get(node, 'NA') for node in nodes]\n",
    "len(nodes)\n",
    "len(nodenames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c70b2dbb-e8e2-4062-af37-0a54baec5bfb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "18251"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(adata.var['GeneStableID']).intersection(flat_col_order))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bc10452b-8720-4077-98b7-4e298850ce36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1418\n",
      "18227\n",
      "1796\n",
      "19876\n",
      "18358\n",
      "1606\n"
     ]
    }
   ],
   "source": [
    "# b_gene_names = set(tenk_b_matrix['GeneName'])\n",
    "\n",
    "\n",
    "b_gene_names = set(flat_col_order)\n",
    "\n",
    "b_gene_names = set(nodenames)\n",
    "\n",
    "tf_names = set(tenk_b_matrix.keys())\n",
    "\n",
    "tfnames = set(adata.obs['TFName'])\n",
    "genes = set(adata.var['GeneName'])\n",
    "\n",
    "\n",
    "print(len(tf_names.intersection(tfnames)))\n",
    "print(len(b_gene_names.intersection(genes)))\n",
    "\n",
    "print(len(tf_names.union(tfnames)))\n",
    "print(len(b_gene_names.union(genes)))\n",
    "\n",
    "print(len(b_gene_names))\n",
    "print(len(tf_names))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "e617dfe6-d4aa-4b36-a501-b4b0848376df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>GeneName</th>\n",
       "      <th>ACADM</th>\n",
       "      <th>CNST</th>\n",
       "      <th>DYRK3</th>\n",
       "      <th>TTF2</th>\n",
       "      <th>GREM2</th>\n",
       "      <th>ARNT</th>\n",
       "      <th>PHACTR4</th>\n",
       "      <th>ATF3</th>\n",
       "      <th>ATP1A4</th>\n",
       "      <th>ATP1B1</th>\n",
       "      <th>...</th>\n",
       "      <th>OR51A4</th>\n",
       "      <th>NPIPB12</th>\n",
       "      <th>IGKV1D-16</th>\n",
       "      <th>UPK3BL2</th>\n",
       "      <th>IGKV1-9</th>\n",
       "      <th>IGHG4</th>\n",
       "      <th>CXorf51A</th>\n",
       "      <th>FAM236C</th>\n",
       "      <th>DEFB130B</th>\n",
       "      <th>TAF11L13</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TFName</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NFYA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARX</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HOXA11</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEOX1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SOX8</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCGF2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF670</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF8</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF229</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NR2E3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1608 rows × 19782 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "GeneName  ACADM  CNST  DYRK3  TTF2  GREM2  ARNT  PHACTR4  ATF3  ATP1A4  \\\n",
       "TFName                                                                   \n",
       "NFYA          0     0      0     0      0   0.0        0   0.0       0   \n",
       "ARX           0     0      0     0      0   0.0        0   0.0       0   \n",
       "HOXA11        0     0      0     0      0   0.0        0   0.0       0   \n",
       "MEOX1         0     0      0     0      0   0.0        0   0.0       0   \n",
       "SOX8          0     0      0     0      0   0.0        0   0.0       0   \n",
       "...         ...   ...    ...   ...    ...   ...      ...   ...     ...   \n",
       "PCGF2         0     0      0     0      0   0.0        0   0.0       0   \n",
       "ZNF670        0     0      0     0      0   0.0        0   0.0       0   \n",
       "ZNF8          0     0      0     0      0   0.0        0   0.0       0   \n",
       "ZNF229        0     0      0     0      0   0.0        0   0.0       0   \n",
       "NR2E3         0     0      0     0      0   0.0        0   0.0       0   \n",
       "\n",
       "GeneName  ATP1B1  ...  OR51A4  NPIPB12  IGKV1D-16  UPK3BL2  IGKV1-9  IGHG4  \\\n",
       "TFName            ...                                                        \n",
       "NFYA           0  ...       0        0          0        0        0      0   \n",
       "ARX            0  ...       0        0          0        0        0      0   \n",
       "HOXA11         0  ...       0        0          0        0        0      0   \n",
       "MEOX1          0  ...       0        0          0        0        0      0   \n",
       "SOX8           0  ...       0        0          0        0        0      0   \n",
       "...          ...  ...     ...      ...        ...      ...      ...    ...   \n",
       "PCGF2          0  ...       0        0          0        0        0      0   \n",
       "ZNF670         0  ...       0        0          0        0        0      0   \n",
       "ZNF8           0  ...       0        0          0        0        0      0   \n",
       "ZNF229         0  ...       0        0          0        0        0      0   \n",
       "NR2E3          0  ...       0        0          0        0        0      0   \n",
       "\n",
       "GeneName  CXorf51A  FAM236C  DEFB130B  TAF11L13  \n",
       "TFName                                           \n",
       "NFYA             0        0         0         0  \n",
       "ARX              0        0         0         0  \n",
       "HOXA11           0        0         0         0  \n",
       "MEOX1            0        0         0         0  \n",
       "SOX8             0        0         0         0  \n",
       "...            ...      ...       ...       ...  \n",
       "PCGF2            0        0         0         0  \n",
       "ZNF670           0        0         0         0  \n",
       "ZNF8             0        0         0         0  \n",
       "ZNF229           0        0         0         0  \n",
       "NR2E3            0        0         0         0  \n",
       "\n",
       "[1608 rows x 19782 columns]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfnames = adata.obs['TFName']\n",
    "genenames = adata.var['GeneName']\n",
    "\n",
    "tenk_aligned = tenk_b_matrix.reindex(index=tfnames, columns=genenames, fill_value=0)\n",
    "five_b_aligned = five_h_b_matrix.reindex(index=tfnames, columns=genenames, fill_value=0)\n",
    "\n",
    "adata.layers['10kp_up_10kb_down'] = tenk_aligned.to_numpy()\n",
    "adata.layers['500bp_up_100bp_down'] = five_b_aligned.to_numpy()\n",
    "\n",
    "tenk_aligned\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "f339144d-d367-4dbe-aef5-1ac1082bbf3d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of non zero interactions for threshold 1.5 and tenk_aligned region: 0.00%\n",
      "number of zero interactions for threshold 1.5 and tenk_aligned region: 1.00%\n"
     ]
    }
   ],
   "source": [
    "print_stats(tenk_aligned, 'tenk_aligned')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "ad79e9b4-a3e6-41db-86f7-533052d9e175",
   "metadata": {},
   "outputs": [],
   "source": [
    "threshold = 1.5\n",
    "\n",
    "tenk_pure_mat = tenk_b_matrix.select_dtypes(include='number')\n",
    "\n",
    "five_h_pure_mat = five_h_b_matrix.select_dtypes(include='number')\n",
    "\n",
    "thresh_tenk_b_matrix = tenk_pure_mat[tenk_pure_mat > threshold]\n",
    "\n",
    "thresh_five_h_b_matrix = five_h_pure_mat[five_h_pure_mat > threshold]\n",
    "\n",
    "binary_tenk_b_matrix = (thresh_tenk_b_matrix > threshold).astype(int)\n",
    "binary_five_h_b_matrix = (thresh_five_h_b_matrix > threshold).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "176dba9a-b77c-4b47-ac83-c2c95c80fd35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of non zero interactions for threshold 1.5 and larger region: 0.75%\n",
      "number of zero interactions for threshold 1.5 and larger region: 0.25%\n",
      "number of non zero interactions for threshold 1.5 and smaller region: 0.21%\n",
      "number of zero interactions for threshold 1.5 and smaller region: 0.79%\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def print_stats(df, label):\n",
    "    \n",
    "    non_zero_count = (df != 0).sum().sum()\n",
    "    zero_count = (df == 0).sum().sum()\n",
    "    total = df.size\n",
    "    \n",
    "    non_zero_percent = non_zero_count / total\n",
    "    zero_percent = zero_count / total\n",
    "\n",
    "\n",
    "    print(f\"number of non zero interactions for threshold {threshold} and {label} region: {non_zero_percent:.2f}%\")\n",
    "    print(f\"number of zero interactions for threshold {threshold} and {label} region: {zero_percent:.2f}%\")\n",
    "\n",
    "df = binary_tenk_b_matrix\n",
    "print_stats(df, 'larger')\n",
    "\n",
    "df = binary_five_h_b_matrix\n",
    "print_stats(df, 'smaller')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "89d1a3e4-643a-4177-b16a-95bc8c965c29",
   "metadata": {},
   "outputs": [],
   "source": [
    "binary_tenk_b_matrix['GeneName'] = tenk_b_matrix['Unnamed: 0']\n",
    "binary_five_h_b_matrix['GeneName'] = five_h_b_matrix['Unnamed: 0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "e0c8d67c-df09-41fc-9811-5b27bd4b6ae0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(27090, 1606)"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.write('/scratch/indikar_root/indikar1/shared_data/HWG/operations/B_matrices.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "abb6685d-4d0e-448e-b70b-7d11329bfc8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1, 1, 1, ..., 0, 0, 0],\n",
       "       [0, 0, 0, ..., 0, 0, 0],\n",
       "       [0, 0, 0, ..., 0, 0, 0],\n",
       "       ...,\n",
       "       [0, 0, 0, ..., 0, 0, 0],\n",
       "       [0, 0, 0, ..., 0, 0, 0],\n",
       "       [0, 0, 0, ..., 0, 0, 0]], shape=(1608, 19782))"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b_prime_2_0 = pd.DataFrame(adata.X, index=adata.obs[''], columns=adata.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e6ed30dc-34e1-4f61-b0e5-1b0b24f9d08e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ordering 18582 genes\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "18582"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from mapping import get_sorted_gene_order\n",
    "\n",
    "flat_col_order = get_sorted_gene_order()\n",
    "\n",
    "nodes = flat_col_order\n",
    "len(nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39b6ddb7-f0a6-4901-a9cc-d130bc702bd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "import networkx as nx\n",
    "\n",
    "\n",
    "G = nx.from_pandas_edgelist(A_matrix_edgelist, source='source', target='target', edge_attr='weight', create_using=nx.DiGraph())\n",
    "\n",
    "adj_matrix = nx.to_scipy_sparse_array(G, nodelist=sorted(G.nodes()), weight='weight', format='csr')\n",
    "\n",
    "adata = AnnData(X=adj_matrix)\n",
    "adata.obs_names = nodes\n",
    "adata.var_names = nodes\n",
    "\n",
    "adata.write('/scratch/indikar_root/indikar1/shared_data/HWG/data/HWG/A_matrix.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "363a4fa7-ff0f-422c-970d-3827ee23dfb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from constants import HWG_BASE_PATH\n",
    "import os\n",
    "\n",
    "from mapping import get_string_protein_gene_map, get_TF_set\n",
    "from constants import HARD_WIRED_GENOME_A_EDGELIST, STRING_EXTRACTED_URL_PATH, HI_UNION_PATH, LIT_BM_PATH\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "04d30076-257c-4f96-b1a5-e78bc31b2dc6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructing Hard Wired Genome A Matrix\n",
      "STRING interactions processed:  13715404\n"
     ]
    }
   ],
   "source": [
    "\n",
    "threshold = 0\n",
    "\n",
    "if not os.path.exists(HWG_BASE_PATH):\n",
    "    os.makedirs(HWG_BASE_PATH)\n",
    "print(\"Constructing Hard Wired Genome A Matrix\")\n",
    "\n",
    "if nodes is None:\n",
    "    nodes = get_sorted_gene_order()\n",
    "\n",
    "# with open(ENSEMBL_PROTEIN_GENE_PATH, \"r\") as protein_gene_file:\n",
    "#     protein_gene_map = {}\n",
    "#     for line in protein_gene_file:\n",
    "#         gene, protein = line.split()\n",
    "#         protein_gene_map[protein] = gene\n",
    "\n",
    "string_protein_gene_map = get_string_protein_gene_map()\n",
    "\n",
    "n = len(nodes)\n",
    "\n",
    "node_idx = {g: i for i, g in enumerate(nodes)}\n",
    "adj = np.zeros((len(nodes), len(nodes)), dtype=int)\n",
    "\n",
    "# adj_df = pd.DataFrame(0, index=nodes, columns=nodes)\n",
    "\n",
    "# initialize edge set\n",
    "str_edges = set()\n",
    "count = 0\n",
    "\n",
    "with open(STRING_EXTRACTED_URL_PATH, \"r\") as f: \n",
    "    string_links = f.readlines()\n",
    "\n",
    "    # skipping header line\n",
    "    for link in string_links[1:]:\n",
    "        count += 1\n",
    "        p1, p2, score = link.strip().split(' ')\n",
    "        p1 = p1.split('.')[1]\n",
    "        p2 = p2.split('.')[1]\n",
    "        score = int(score)\n",
    "        g1 = string_protein_gene_map.get(p1, None)\n",
    "        g2 = string_protein_gene_map.get(p2, None)\n",
    "        # if score >= threshold:\n",
    "        \n",
    "        # if g1 in nodes and g2 in nodes:\n",
    "        if g1 is not None and g2 is not None and score >= threshold:\n",
    "            str_edges.add((g1, g2, score))\n",
    "print(\"STRING interactions processed: \", len(str_edges))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "da743ba9-d229-4285-ae3c-12fd7e11f817",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HI-UNION interactions processed:  64006\n",
      "LIT-BM interactions processed:  12578\n",
      "processed 1639 Transcription Factors\n",
      "['DUX1_HUMAN', 'DUX3_HUMAN', 'ENSG00000001167', 'ENSG00000004848', 'ENSG00000005073', 'ENSG00000005102', 'ENSG00000005513', 'ENSG00000005801', 'ENSG00000005889', 'ENSG00000006047']\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "unsupported operand type(s) for +: 'set' and 'set'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mTypeError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[31]\u001b[39m\u001b[32m, line 23\u001b[39m\n\u001b[32m     18\u001b[39m \u001b[38;5;66;03m# for g1, g2, inter in edges:\u001b[39;00m\n\u001b[32m     19\u001b[39m \u001b[38;5;66;03m#     if g1 in node_idx and g2 in node_idx:\u001b[39;00m\n\u001b[32m     20\u001b[39m \u001b[38;5;66;03m#         i, j = node_idx[g1], node_idx[g2]\u001b[39;00m\n\u001b[32m     21\u001b[39m \u001b[38;5;66;03m#         adj[i, j] = 1\u001b[39;00m\n\u001b[32m     22\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mpandas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mpd\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m23\u001b[39m edges = \u001b[43mhuri_edges\u001b[49m\u001b[43m \u001b[49m\u001b[43m+\u001b[49m\u001b[43m \u001b[49m\u001b[43mstr_edges\u001b[49m\n\u001b[32m     24\u001b[39m edgelist = \u001b[38;5;28mlist\u001b[39m(edges)\n\u001b[32m     25\u001b[39m edgelist_df = pd.DataFrame(edgelist, columns=[\u001b[33m'\u001b[39m\u001b[33msource\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mtarget\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mweight\u001b[39m\u001b[33m'\u001b[39m])\n",
      "\u001b[31mTypeError\u001b[39m: unsupported operand type(s) for +: 'set' and 'set'"
     ]
    }
   ],
   "source": [
    "\n",
    "huri_edges = set()\n",
    "with open(HI_UNION_PATH, \"r\") as huri_links:\n",
    "    for link in huri_links:\n",
    "        g1, g2 = link.strip().split()\n",
    "        huri_edges.add((g1, g2, 1))\n",
    "print(\"HI-UNION interactions processed: \", len(huri_edges))\n",
    "huri_count = len(huri_edges)\n",
    "\n",
    "with open(LIT_BM_PATH, \"r\") as lit_links:\n",
    "    for link in lit_links:\n",
    "        g1, g2 = link.strip().split()\n",
    "        huri_edges.add((g1, g2, 1))\n",
    "print(\"LIT-BM interactions processed: \", len(huri_edges) - huri_count)\n",
    "\n",
    "tflist = get_TF_set()\n",
    "print(tflist[:10])\n",
    "\n",
    "# for g1, g2, inter in edges:\n",
    "#     if g1 in node_idx and g2 in node_idx:\n",
    "#         i, j = node_idx[g1], node_idx[g2]\n",
    "#         adj[i, j] = 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e4e92635-376c-4c36-9685-17446e2cf22d",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'str_edgelest_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[33]\u001b[39m\u001b[32m, line 28\u001b[39m\n\u001b[32m     26\u001b[39m \u001b[38;5;66;03m# STRING edges\u001b[39;00m\n\u001b[32m     27\u001b[39m str_edgelist_df = pd.DataFrame(str_edges, columns=[\u001b[33m'\u001b[39m\u001b[33msource\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mtarget\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mweight\u001b[39m\u001b[33m'\u001b[39m])\n\u001b[32m---> \u001b[39m\u001b[32m28\u001b[39m G2 = nx.from_pandas_edgelist(\u001b[43mstr_edgelest_df\u001b[49m, source=\u001b[33m'\u001b[39m\u001b[33msource\u001b[39m\u001b[33m'\u001b[39m, target=\u001b[33m'\u001b[39m\u001b[33mtarget\u001b[39m\u001b[33m'\u001b[39m, edge_attr=\u001b[33m'\u001b[39m\u001b[33mweight\u001b[39m\u001b[33m'\u001b[39m, create_using=nx.DiGraph())\n\u001b[32m     29\u001b[39m G2.add_nodes_from(nodes)\n\u001b[32m     30\u001b[39m adj_matrix = nx.to_scipy_sparse_array(G2, nodelist=nodes, weight=\u001b[33m'\u001b[39m\u001b[33mweight\u001b[39m\u001b[33m'\u001b[39m, \u001b[38;5;28mformat\u001b[39m=\u001b[33m'\u001b[39m\u001b[33mcsr\u001b[39m\u001b[33m'\u001b[39m)\n",
      "\u001b[31mNameError\u001b[39m: name 'str_edgelest_df' is not defined"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import networkx as nx\n",
    "edges = huri_edges | str_edges\n",
    "edgelist = list(edges)\n",
    "edgelist_df = pd.DataFrame(edgelist, columns=['source', 'target', 'weight'])\n",
    "\n",
    "\n",
    "# adj_df = pd.DataFrame(adj, index=nodes, columns=nodes)\n",
    "\n",
    "save_path = HARD_WIRED_GENOME_A_EDGELIST.strip('.csv') + f\"_{threshold}.csv\"\n",
    "# edgelist_df.to_csv(save_path)\n",
    "A_matrix_edgelist = edgelist_df\n",
    "\n",
    "\n",
    "\n",
    "G = nx.from_pandas_edgelist(A_matrix_edgelist, source='source', target='target', edge_attr='weight', create_using=nx.DiGraph())\n",
    "G.add_nodes_from(nodes)\n",
    "nodes = sorted(G.nodes())\n",
    "adj_matrix = nx.to_scipy_sparse_array(G, nodelist=nodes, weight='weight', format='csr')\n",
    "\n",
    "\n",
    "adata = AnnData(X=adj_matrix)\n",
    "adata.obs_names = nodes\n",
    "adata.var_names = nodes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "59090cf8-e571-4ba9-9cbf-ecb900d9f004",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "import networkx as nx\n",
    "adata = anndata.read_h5ad('/scratch/indikar_root/indikar1/shared_data/HWG/data/HWG/A_matrix.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "200c84c6-d8b0-41aa-a58d-6c3a345c3d6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# STRING edges\n",
    "# str_edgelist_df = pd.DataFrame(str_edges, columns=['source', 'target', 'weight'])\n",
    "# G2 = nx.from_pandas_edgelist(str_edgelist_df, source='source', target='target', edge_attr='weight', create_using=nx.DiGraph())\n",
    "G2.add_nodes_from(nodes)\n",
    "\n",
    "nodes = sorted(G2.nodes())\n",
    "adj_matrix = nx.to_scipy_sparse_array(G2, nodelist=nodes, weight='weight', format='csr')\n",
    "\n",
    "adata.layers['STRING'] = adj_matrix\n",
    "\n",
    "\n",
    "# HURI edges\n",
    "# huri_edgelist_df = pd.DataFrame(huri_edges, columns=['source', 'target', 'weight'])\n",
    "# G3 = nx.from_pandas_edgelist(huri_edgelist_df, source='source', target='target', edge_attr='weight', create_using=nx.DiGraph())\n",
    "# G3.add_nodes_from(nodes)\n",
    "# adj_matrix = nx.to_scipy_sparse_array(G3, nodelist=nodes, weight='weight', format='csr')\n",
    "\n",
    "# adata.layers['HURI'] = adj_matrix\n",
    "\n",
    "# print(\"Constructing Hard Wired Genome B Matrix\")\n",
    "# B_matrix_edgelist = A_matrix_edgelist[\n",
    "# A_matrix_edgelist['source'].isin(flat_col_order) & A_matrix_edgelist['target'].isin(tflist)\n",
    "# ]\n",
    "# print(f'processed edges {len(B_matrix_edgelist)}')\n",
    "\n",
    "# print(\"Constructing Hard Wired Genome C Matrix\")\n",
    "# C_matrix_edgelist = A_matrix_edgelist[\n",
    "#     A_matrix_edgelist['source'].isin(tflist) & A_matrix_edgelist['target'].isin(tflist)\n",
    "# ]\n",
    "# print(f'processed edges {len(C_matrix_edgelist)}')\n",
    "\n",
    "\n",
    "# save_path = HARD_WIRED_GENOME_B_EDGELIST.strip('.csv') + f\"_{threshold}.csv\"\n",
    "# B_matrix_edgelist.to_csv(save_path, index=False)\n",
    "\n",
    "# save_path = HARD_WIRED_GENOME_C_EDGELIST.strip('.csv') + f\"_{threshold}.csv\"\n",
    "# C_matrix_edgelist.to_csv(save_path, index=False)\n",
    "\n",
    "# print(f\"Saved Hard Wired Genome to {save_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "0c3d6caf-ab4c-48ac-b8e2-460287bb84f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write('/scratch/indikar_root/indikar1/shared_data/HWG/data/HWG/A_matrix.h5ad')\n",
    "\n",
    "# adata = anndata.read_h5ad('/scratch/indikar_root/indikar1/shared_data/HWG/data/HWG/A_matrix.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70f35c24-b23d-428d-b61d-ae0faaf864bf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "HardWiredGenome",
   "language": "python",
   "name": "hardwiredgenome"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
